// @bitgo/account-lib/dist/src/coin/trx/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PermissionType = exports.ContractType = void 0;
/**
 * Contract type describes a class of contract in a Transaction. This structure is used to identify what the Contract is.
 */
var ContractType;
(function (ContractType) {
    /**
     * This is a transfer of TRX contract.
     */
    ContractType[ContractType["Transfer"] = 0] = "Transfer";
    /**
     * This is the multi-sig initialization contract.
     */
    ContractType[ContractType["AccountPermissionUpdate"] = 1] = "AccountPermissionUpdate";
    /**
     * This is a smart contract type.
     */
    ContractType[ContractType["TriggerSmartContract"] = 2] = "TriggerSmartContract";
})(ContractType = exports.ContractType || (exports.ContractType = {}));
var PermissionType;
(function (PermissionType) {
    PermissionType[PermissionType["Owner"] = 0] = "Owner";
    PermissionType[PermissionType["Witness"] = 1] = "Witness";
    PermissionType[PermissionType["Active"] = 2] = "Active";
})(PermissionType = exports.PermissionType || (exports.PermissionType = {}));
