// @bitgo/account-lib/dist/src/coin/trx/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.Interface = exports.Enum = exports.ContractCallBuilder = exports.WrappedBuilder = exports.Transaction = exports.Address = exports.KeyPair = void 0;
var keyPair_1 = require("./keyPair");
Object.defineProperty(exports, "KeyPair", { enumerable: true, get: function () { return keyPair_1.KeyPair; } });
var address_1 = require("./address");
Object.defineProperty(exports, "Address", { enumerable: true, get: function () { return address_1.Address; } });
var transaction_1 = require("./transaction");
Object.defineProperty(exports, "Transaction", { enumerable: true, get: function () { return transaction_1.Transaction; } });
var wrappedBuilder_1 = require("./wrappedBuilder");
Object.defineProperty(exports, "WrappedBuilder", { enumerable: true, get: function () { return wrappedBuilder_1.WrappedBuilder; } });
var contractCallBuilder_1 = require("./contractCallBuilder");
Object.defineProperty(exports, "ContractCallBuilder", { enumerable: true, get: function () { return contractCallBuilder_1.ContractCallBuilder; } });
var Enum = __importStar(require("./enum"));
exports.Enum = Enum;
var Interface = __importStar(require("./iface"));
exports.Interface = Interface;
var Utils = __importStar(require("./utils"));
exports.Utils = Utils;
