// @bitgo/account-lib/dist/src/coin/stx/contractBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractBuilder = void 0;
var transactions_1 = require("@stacks/transactions");
var errors_1 = require("../baseCoin/errors");
var utils_1 = require("./utils");
var _1 = require(".");
var constants_1 = require("./constants");
var abstractContractBuilder_1 = require("./abstractContractBuilder");
var ContractBuilder = /** @class */ (function (_super) {
    __extends(ContractBuilder, _super);
    function ContractBuilder(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    ContractBuilder.prototype.initBuilder = function (tx) {
        _super.prototype.initBuilder.call(this, tx);
        // Retro compatibility, checks parameters.
        this.contractAddress(this._contractAddress);
        this.contractName(this._contractName);
        this.functionName(this._functionName);
        this.functionArgs(this._functionArgs);
    };
    // region Contract fields
    /**
     * Set the contract address
     *
     * @param {string} address the address deployed the contract
     * @returns {ContractBuilder} the builder with the new parameter set
     */
    ContractBuilder.prototype.contractAddress = function (address) {
        if (!utils_1.isValidAddress(address)) {
            throw new errors_1.InvalidParameterValueError('Invalid address');
        }
        if (!_1.Utils.isValidContractAddress(address, this._coinConfig.network)) {
            throw new errors_1.InvalidParameterValueError('Invalid contract address');
        }
        this._contractAddress = address;
        return this;
    };
    /**
     * Set the contract name
     *
     * @param {string} name name of contract
     * @returns {ContractBuilder} the builder with the new parameter set
     */
    ContractBuilder.prototype.contractName = function (name) {
        if (name.length === 0) {
            throw new errors_1.InvalidParameterValueError('Invalid name');
        }
        if (name !== constants_1.CONTRACT_NAME_STAKING && name !== constants_1.CONTRACT_NAME_SENDMANY) {
            throw new errors_1.InvalidParameterValueError('Only pox and send-many-memo contracts supported');
        }
        this._contractName = name;
        return this;
    };
    /**
     * Set the function name in contract
     *
     * @param {string} name name of function
     * @returns {ContractBuilder} the builder with the new parameter set
     */
    ContractBuilder.prototype.functionName = function (name) {
        if (name.length === 0) {
            throw new errors_1.InvalidParameterValueError('Invalid name');
        }
        if (!_1.Utils.isValidContractFunctionName(name)) {
            throw new errors_1.InvalidParameterValueError(name + " is not supported contract function name");
        }
        this._functionName = name;
        return this;
    };
    ContractBuilder.prototype.functionArgs = function (args) {
        var _this = this;
        this._functionArgs = args.map(function (arg) {
            if (!transactions_1.ClarityType[arg.type]) {
                return _this.parseCv(arg);
            }
            else {
                // got direct clarity value after deserialization in fromImplementation
                return arg;
            }
        });
        return this;
    };
    ContractBuilder.prototype.parseCv = function (arg) {
        var _this = this;
        switch (arg.type) {
            case 'optional':
                if (arg.val === undefined) {
                    return transactions_1.noneCV();
                }
                else {
                    return transactions_1.someCV(this.parseCv(arg.val));
                }
            case 'tuple':
                if (arg.val instanceof Array) {
                    var data_1 = {};
                    arg.val.forEach(function (a) {
                        data_1[a.key] = _this.parseCv({ type: a.type, val: a.val });
                    });
                    return transactions_1.tupleCV(data_1);
                }
                throw new errors_1.InvalidParameterValueError('tuple require Array val');
            case 'buffer':
                if (arg.val instanceof Buffer) {
                    return transactions_1.bufferCV(arg.val);
                }
                var nval = Number(arg.val);
                if (nval) {
                    return transactions_1.bufferCV(Buffer.of(nval));
                }
                return transactions_1.bufferCVFromString(arg.val);
            default:
                return transactions_1.encodeClarityValue(arg.type, arg.val);
        }
    };
    return ContractBuilder;
}(abstractContractBuilder_1.AbstractContractBuilder));
exports.ContractBuilder = ContractBuilder;
