// @bitgo/account-lib/dist/src/coin/stx/transactionBuilderFactory.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
var transactions_1 = require("@stacks/transactions");
var baseCoin_1 = require("../baseCoin");
var errors_1 = require("../baseCoin/errors");
var transferBuilder_1 = require("./transferBuilder");
var transaction_1 = require("./transaction");
var contractBuilder_1 = require("./contractBuilder");
var _1 = require(".");
var sendmanyBuilder_1 = require("./sendmanyBuilder");
var TransactionBuilderFactory = /** @class */ (function (_super) {
    __extends(TransactionBuilderFactory, _super);
    function TransactionBuilderFactory(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    /** @inheritdoc */
    TransactionBuilderFactory.prototype.from = function (raw) {
        this.validateRawTransaction(raw);
        try {
            var tx = this.parseTransaction(raw);
            switch (tx.stxTransaction.payload.payloadType) {
                case transactions_1.PayloadType.TokenTransfer:
                    return this.getTransferBuilder(tx);
                case transactions_1.PayloadType.ContractCall:
                    if (sendmanyBuilder_1.SendmanyBuilder.isValidContractCall(this._coinConfig, tx.stxTransaction.payload)) {
                        return this.getSendmanyBuilder(tx);
                    }
                    return this.getContractBuilder(tx);
                default:
                    throw new errors_1.InvalidTransactionError('Invalid transaction');
            }
        }
        catch (e) {
            console.error(e);
            throw new errors_1.ParseTransactionError('There was an error parsing the raw transaction');
        }
    };
    TransactionBuilderFactory.prototype.parseTransaction = function (rawTransaction) {
        var tx = new transaction_1.Transaction(this._coinConfig);
        var stackstransaction = transactions_1.deserializeTransaction(transactions_1.BufferReader.fromBuffer(Buffer.from(_1.Utils.removeHexPrefix(rawTransaction), 'hex')));
        tx.stxTransaction = stackstransaction;
        return tx;
    };
    /** @inheritdoc */
    TransactionBuilderFactory.prototype.getWalletInitializationBuilder = function (tx) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    /** @inheritdoc */
    TransactionBuilderFactory.prototype.getTransferBuilder = function (tx) {
        return TransactionBuilderFactory.initializeBuilder(new transferBuilder_1.TransferBuilder(this._coinConfig), tx);
    };
    TransactionBuilderFactory.prototype.getContractBuilder = function (tx) {
        return TransactionBuilderFactory.initializeBuilder(new contractBuilder_1.ContractBuilder(this._coinConfig), tx);
    };
    TransactionBuilderFactory.prototype.getSendmanyBuilder = function (tx) {
        return TransactionBuilderFactory.initializeBuilder(new sendmanyBuilder_1.SendmanyBuilder(this._coinConfig), tx);
    };
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    TransactionBuilderFactory.initializeBuilder = function (builder, tx) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    };
    /** @inheritdoc */
    TransactionBuilderFactory.prototype.validateRawTransaction = function (rawTransaction) {
        if (!rawTransaction) {
            throw new errors_1.InvalidTransactionError('Raw transaction is empty');
        }
        if (!_1.Utils.isValidRawTransaction(rawTransaction)) {
            throw new errors_1.ParseTransactionError('Invalid raw transaction');
        }
    };
    return TransactionBuilderFactory;
}(baseCoin_1.BaseTransactionBuilderFactory));
exports.TransactionBuilderFactory = TransactionBuilderFactory;
