// @bitgo/account-lib/dist/src/coin/algo/keyRegistrationBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyRegistrationBuilder = void 0;
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var algosdk_1 = __importDefault(require("algosdk"));
var baseCoin_1 = require("../baseCoin");
var errors_1 = require("../baseCoin/errors");
var transactionBuilder_1 = require("./transactionBuilder");
var txnSchema_1 = require("./txnSchema");
var utils_1 = __importDefault(require("./utils"));
var KeyRegistrationBuilder = /** @class */ (function (_super) {
    __extends(KeyRegistrationBuilder, _super);
    function KeyRegistrationBuilder(coinConfig) {
        return _super.call(this, coinConfig) || this;
    }
    /**
     * Sets the vote key
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * @param {string} key The root participation public key. See Generate a Participation Key to learn more.
     * https://developer.algorand.org/docs/reference/transactions/#key-registration-transaction
     */
    KeyRegistrationBuilder.prototype.voteKey = function (key) {
        this._voteKey = key;
        return this;
    };
    /**
     * Sets the selection key
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * @param {string} key The VRF public key for the account.
     * https://developer.algorand.org/docs/reference/transactions/#key-registration-transaction
     */
    KeyRegistrationBuilder.prototype.selectionKey = function (key) {
        this._selectionKey = key;
        return this;
    };
    /**
     * Sets the stateProof key
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * @param {string} key The stateproof key. See consensus for more information.
     * https://developer.algorand.org/docs/get-details/algorand_consensus/?from_query=state#state-proof-keys
     */
    KeyRegistrationBuilder.prototype.stateProofKey = function (key) {
        utils_1.default.validateBase64(key);
        this._stateProofKey = key;
        return this;
    };
    /**
     *Sets the vote first round
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * @param {number} round The first round that the participation key is valid. Not to be confused with the FirstValid round of the keyreg transaction.
     * https://developer.algorand.org/docs/reference/transactions/#key-registration-transaction
     */
    KeyRegistrationBuilder.prototype.voteFirst = function (round) {
        this.validateValue(new bignumber_js_1.default(round));
        this._voteFirst = round;
        return this;
    };
    /**
     * Sets the vote last round
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * A recommended range is 3,000,000 rounds.
     *
     * @param {number} round No theoretical limit.
     * https://developer.algorand.org/docs/run-a-node/participate/generate_keys/
     */
    KeyRegistrationBuilder.prototype.voteLast = function (round) {
        this.validateValue(new bignumber_js_1.default(round));
        this._voteLast = round;
        return this;
    };
    /**
     * Sets the vote key dilution
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * Defaults to 10,000
     *
     * @param {number} size [10000]. To reduce the size of the participation key, set the key dilution value to roughly the square root of the range that the partkey is valid for.
     * https://developer.algorand.org/docs/run-a-node/participate/generate_keys/#generate-the-participation-key-with-goal
     * @param size
     */
    KeyRegistrationBuilder.prototype.voteKeyDilution = function (size) {
        if (size === void 0) { size = 10000; }
        this.validateValue(new bignumber_js_1.default(size));
        this._voteKeyDilution = size;
        return this;
    };
    /**
     * Sets the non participation flag
     *
     * @returns {KeyRegistrationBuilder} This Key Registration builder.
     *
     * @param {boolean} nonParticipation All new Algorand accounts are participating by default.
     * This means that they earn rewards. Mark an account nonparticipating by setting this value to true and this account
     * will no longer earn rewards NEVER.
     * https://developer.algorand.org/docs/reference/transactions/#key-registration-transaction
     */
    KeyRegistrationBuilder.prototype.nonParticipation = function (nonParticipation) {
        this._nonParticipation = nonParticipation;
        return this;
    };
    KeyRegistrationBuilder.prototype.buildAlgoTxn = function () {
        return this.isOfflineKeyRegAccountLibTransaction()
            ? this.buildOfflineKeyRegTransaction()
            : this.isNonParticipationKeyRegAccountLibTransaction()
                ? this.buildNonParticipationKeyRegTransaction()
                : this.buildOnlineKeyRegTransaction();
    };
    KeyRegistrationBuilder.prototype.buildOfflineKeyRegTransaction = function () {
        return algosdk_1.default.makeKeyRegistrationTxnWithSuggestedParams(this._sender, this._note, undefined, // voteKey param
        undefined, // selectionKey param
        undefined, // voteFirst param
        undefined, // voteLast param
        undefined, // voteKeyDilution param
        this.suggestedParams);
    };
    KeyRegistrationBuilder.prototype.buildOnlineKeyRegTransaction = function () {
        return algosdk_1.default.makeKeyRegistrationTxnWithSuggestedParams(this._sender, this._note, this._voteKey, this._selectionKey, this._voteFirst, this._voteLast, this._voteKeyDilution, this.suggestedParams, undefined, // reKeyTo param
        undefined, // nonParticipation param
        this._stateProofKey);
    };
    KeyRegistrationBuilder.prototype.buildNonParticipationKeyRegTransaction = function () {
        return algosdk_1.default.makeKeyRegistrationTxnWithSuggestedParams(this._sender, this._note, undefined, // voteKey param
        undefined, // selectionKey param
        undefined, // voteFirst param
        undefined, // voteLast param
        undefined, // voteKeyDilution param
        this.suggestedParams, this._reKeyTo, true);
    };
    Object.defineProperty(KeyRegistrationBuilder.prototype, "transactionType", {
        get: function () {
            return baseCoin_1.TransactionType.WalletInitialization;
        },
        enumerable: false,
        configurable: true
    });
    /** @inheritdoc */
    KeyRegistrationBuilder.prototype.fromImplementation = function (rawTransaction) {
        var tx = _super.prototype.fromImplementation.call(this, rawTransaction);
        var algoTxn = tx.getAlgoTransaction();
        if (algoTxn) {
            if (this.isNonParticipationKeyRegAlgoSDKTransaction(algoTxn)) {
                this.nonParticipation(!!algoTxn.nonParticipation);
            }
            else if (this.isOnlineKeyRegAlgoSDKTransaction(algoTxn)) {
                this.voteKey(algoTxn.voteKey.toString('base64'));
                this.selectionKey(algoTxn.selectionKey.toString('base64'));
                this.voteFirst(algoTxn.voteFirst);
                this.voteLast(algoTxn.voteLast);
                this.voteKeyDilution(algoTxn.voteKeyDilution);
                if (algoTxn.stateProofKey) {
                    this.stateProofKey(algoTxn.stateProofKey.toString('base64'));
                }
            }
        }
        return tx;
    };
    /** @inheritdoc */
    KeyRegistrationBuilder.prototype.validateRawTransaction = function (rawTransaction) {
        var decodeTxn = utils_1.default.decodeAlgoTxn(rawTransaction);
        var algoTxn = decodeTxn.txn;
        if (algoTxn.type !== algosdk_1.default.TransactionType.keyreg) {
            throw new errors_1.InvalidTransactionError("Invalid Transaction Type: " + algoTxn.type + ". Expected " + algosdk_1.default.TransactionType.keyreg);
        }
        if (this.isOnlineKeyRegAlgoSDKTransaction(algoTxn)) {
            this.validateFields(algoTxn.voteKey.toString('base64'), algoTxn.selectionKey.toString('base64'), algoTxn.voteFirst, algoTxn.voteLast, algoTxn.voteKeyDilution, algoTxn.stateProofKey && algoTxn.stateProofKey.toString('base64'));
        }
    };
    KeyRegistrationBuilder.prototype.isNonParticipationKeyRegAlgoSDKTransaction = function (algoTxn) {
        return !!algoTxn.nonParticipation;
    };
    KeyRegistrationBuilder.prototype.isNonParticipationKeyRegAccountLibTransaction = function () {
        return !!this._nonParticipation;
    };
    KeyRegistrationBuilder.prototype.isOnlineKeyRegAlgoSDKTransaction = function (algoTxn) {
        return !(this.isOfflineKeyRegAlgoSDKTransaction(algoTxn) || this.isNonParticipationKeyRegAlgoSDKTransaction(algoTxn));
    };
    KeyRegistrationBuilder.prototype.isOnlineKeyRegAccountLibTransaction = function () {
        return !(this.isOfflineKeyRegAccountLibTransaction() || this.isNonParticipationKeyRegAccountLibTransaction());
    };
    KeyRegistrationBuilder.prototype.isOfflineKeyRegAlgoSDKTransaction = function (algoTxn) {
        return (!algoTxn.voteKey &&
            !algoTxn.selectionKey &&
            !algoTxn.voteFirst &&
            !algoTxn.voteLast &&
            !algoTxn.voteKeyDilution &&
            !algoTxn.stateProofKey &&
            !algoTxn.nonParticipation);
    };
    KeyRegistrationBuilder.prototype.isOfflineKeyRegAccountLibTransaction = function () {
        return (!this._voteKey &&
            !this._selectionKey &&
            !this._voteFirst &&
            !this._voteLast &&
            !this._voteKeyDilution &&
            !this._stateProofKey &&
            !this._nonParticipation);
    };
    /** @inheritdoc */
    KeyRegistrationBuilder.prototype.validateTransaction = function (transaction) {
        _super.prototype.validateTransaction.call(this, transaction);
        if (this.isOnlineKeyRegAccountLibTransaction()) {
            // invalid offline will reach here
            this.validateFields(this._voteKey, this._selectionKey, this._voteFirst, this._voteLast, this._voteKeyDilution, this._stateProofKey);
        }
        else {
            // offline or nonparticipation transaction
            if (this._voteKey ||
                this._selectionKey ||
                this._voteFirst ||
                this._voteLast ||
                this._voteKeyDilution ||
                this._stateProofKey) {
                throw new errors_1.InvalidTransactionError('VoteKey, SelectionKey, VoteFirst, VoteLast, VoteKeyDilution, StateProofKey fields cannot be set when offline or nonparticipation is set');
            }
        }
    };
    KeyRegistrationBuilder.prototype.validateFields = function (voteKey, selectionKey, voteFirst, voteLast, voteKeyDilution, stateProofKey) {
        var validationResult = txnSchema_1.KeyRegTxnSchema.validate({
            voteKey: voteKey,
            selectionKey: selectionKey,
            voteFirst: voteFirst,
            voteLast: voteLast,
            voteKeyDilution: voteKeyDilution,
            stateProofKey: stateProofKey,
        });
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("Transaction validation failed: " + validationResult.error.message);
        }
    };
    return KeyRegistrationBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.KeyRegistrationBuilder = KeyRegistrationBuilder;
