// @bitgo/account-lib/dist/src/coin/algo/transactionBuilderFactory.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
var algosdk_1 = __importDefault(require("algosdk"));
var errors_1 = require("../baseCoin/errors");
var baseCoin_1 = require("../baseCoin");
var keyRegistrationBuilder_1 = require("./keyRegistrationBuilder");
var transferBuilder_1 = require("./transferBuilder");
var assetTransferBuilder_1 = require("./assetTransferBuilder");
var utils_1 = __importDefault(require("./utils"));
var TransactionBuilderFactory = /** @class */ (function (_super) {
    __extends(TransactionBuilderFactory, _super);
    function TransactionBuilderFactory(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    TransactionBuilderFactory.prototype.getTransferBuilder = function () {
        return new transferBuilder_1.TransferBuilder(this._coinConfig);
    };
    TransactionBuilderFactory.prototype.getAssetTransferBuilder = function () {
        return new assetTransferBuilder_1.AssetTransferBuilder(this._coinConfig);
    };
    TransactionBuilderFactory.prototype.getKeyRegistrationBuilder = function () {
        return new keyRegistrationBuilder_1.KeyRegistrationBuilder(this._coinConfig);
    };
    TransactionBuilderFactory.prototype.from = function (rawTxn) {
        var builder = this.getBuilder(rawTxn);
        builder.from(rawTxn);
        return builder;
    };
    TransactionBuilderFactory.prototype.getBuilder = function (rawTxn) {
        var decodeTxn = utils_1.default.decodeAlgoTxn(rawTxn);
        var algoTxn = decodeTxn.txn;
        if (algoTxn.type == algosdk_1.default.TransactionType.keyreg) {
            return this.getWalletInitializationBuilder();
        }
        else if (algoTxn.type == algosdk_1.default.TransactionType.pay) {
            return this.getTransferBuilder();
        }
        else if (algoTxn.type == algosdk_1.default.TransactionType.axfer) {
            return this.getAssetTransferBuilder();
        }
        else {
            throw new errors_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
    };
    TransactionBuilderFactory.prototype.getWalletInitializationBuilder = function () {
        return new keyRegistrationBuilder_1.KeyRegistrationBuilder(this._coinConfig);
    };
    return TransactionBuilderFactory;
}(baseCoin_1.BaseTransactionBuilderFactory));
exports.TransactionBuilderFactory = TransactionBuilderFactory;
