// @bitgo/account-lib/dist/src/coin/algo/transferBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var algosdk_1 = __importDefault(require("algosdk"));
var errors_1 = require("../baseCoin/errors");
var baseCoin_1 = require("../baseCoin");
var transactionBuilder_1 = require("./transactionBuilder");
var txnSchema_1 = require("./txnSchema");
var utils_1 = __importDefault(require("./utils"));
var TransferBuilder = /** @class */ (function (_super) {
    __extends(TransferBuilder, _super);
    function TransferBuilder(coinConfig) {
        return _super.call(this, coinConfig) || this;
    }
    TransferBuilder.prototype.buildAlgoTxn = function () {
        return algosdk_1.default.makePaymentTxnWithSuggestedParams(this._sender, this._to, this._amount, this._closeRemainderTo, this._note, this.suggestedParams, this._reKeyTo);
    };
    Object.defineProperty(TransferBuilder.prototype, "transactionType", {
        get: function () {
            return baseCoin_1.TransactionType.Send;
        },
        enumerable: false,
        configurable: true
    });
    TransferBuilder.prototype.fromImplementation = function (rawTransaction) {
        var tx = _super.prototype.fromImplementation.call(this, rawTransaction);
        var algoTx = tx.getAlgoTransaction();
        if (algoTx) {
            this.amount(algoTx.amount || 0);
            this.to({ address: algosdk_1.default.encodeAddress(algoTx.to.publicKey) });
            if (algoTx.closeRemainderTo) {
                this.closeRemainderTo({ address: algosdk_1.default.encodeAddress(algoTx.closeRemainderTo.publicKey) });
            }
        }
        return tx;
    };
    TransferBuilder.prototype.validateRawTransaction = function (rawTransaction) {
        var algoTxn = utils_1.default.decodeAlgoTxn(rawTransaction).txn;
        if (algoTxn.type !== algosdk_1.default.TransactionType.pay) {
            throw new errors_1.InvalidTransactionError("Invalid Transaction Type: " + algoTxn.type + ". Expected " + algosdk_1.default.TransactionType.pay);
        }
    };
    TransferBuilder.prototype.validateTransaction = function (transaction) {
        _super.prototype.validateTransaction.call(this, transaction);
        var validationResult = txnSchema_1.TransferTransactionSchema.validate({
            to: this._to,
            amount: this._amount,
            closeRemainderTo: this._closeRemainderTo,
        });
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("Transaction validation failed: " + validationResult.error.message);
        }
    };
    /**
     * Sets the payment receiver.
     *
     * @param {BaseAddress} to The receiver account
     * @returns {TransferBuilder} This transfer builder.
     *
     */
    TransferBuilder.prototype.to = function (to) {
        this.validateAddress(to);
        this._to = to.address;
        return this;
    };
    /**
     * Sets the amount of payment.
     *
     * @param {number} amount The amount of payment.
     * @returns {TransferBuilder} This transfer builder.
     */
    TransferBuilder.prototype.amount = function (amount) {
        this.validateValue(new bignumber_js_1.default(amount.toString()));
        this._amount = amount;
        return this;
    };
    /**
     * Sets address to transfer remainder amount on closing.
     *
     * @param {BaseAddress} closeRemainderTo The address for receiving remainder amount
     * @returns {TransferBuilder} This transfer builder.
     *
     */
    TransferBuilder.prototype.closeRemainderTo = function (closeRemainderTo) {
        this.validateAddress(closeRemainderTo);
        this._closeRemainderTo = closeRemainderTo.address;
        return this;
    };
    return TransferBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.TransferBuilder = TransferBuilder;
