// @bitgo/account-lib/dist/src/coin/algo/txnSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetToggleTxnSchema = exports.AssetTransferTxnSchema = exports.KeyRegTxnSchema = exports.TransferTransactionSchema = exports.BaseTransactionSchema = void 0;
var joi_1 = __importDefault(require("joi"));
var errors_1 = require("../baseCoin/errors");
var utils_1 = __importDefault(require("./utils"));
var addressSchema = joi_1.default.string().custom(function (addr) { return utils_1.default.isValidAddress(addr); });
exports.BaseTransactionSchema = joi_1.default
    .object({
    fee: joi_1.default.number().required(),
    firstRound: joi_1.default.number().positive().required(),
    genesisHash: joi_1.default.string().base64().required(),
    lastRound: joi_1.default.number().positive().required(),
    sender: addressSchema.required(),
    genesisId: joi_1.default.string().optional(),
    lease: joi_1.default.optional(),
    note: joi_1.default.optional(),
    reKeyTo: joi_1.default
        .string()
        .custom(function (addr) { return utils_1.default.isValidAddress(addr); })
        .optional(),
})
    .custom(function (obj) {
    var firstRound = obj.firstRound;
    var lastRound = obj.lastRound;
    if (firstRound < lastRound) {
        return obj;
    }
    throw new Error('lastRound cannot be greater than or equal to firstRound');
});
exports.TransferTransactionSchema = joi_1.default.object({
    amount: joi_1.default.custom(function (val) { return typeof val === 'number' || typeof val === 'bigint'; }).required(),
    to: joi_1.default
        .string()
        .custom(function (addr) { return utils_1.default.isValidAddress(addr); })
        .required(),
    closeRemainderTo: joi_1.default.string().optional(),
});
exports.KeyRegTxnSchema = joi_1.default
    .object({
    voteKey: addressSchema.required(),
    selectionKey: addressSchema.required(),
    voteFirst: joi_1.default.number().positive().required(),
    voteLast: joi_1.default.number().positive().required(),
    voteKeyDilution: joi_1.default.number().positive().required(),
    stateProofKey: addressSchema.optional(),
})
    .custom(function (obj) {
    var voteFirst = obj.voteFirst;
    var voteLast = obj.voteLast;
    if (voteFirst > voteLast) {
        throw new errors_1.InvalidTransactionError('VoteKey last round must be greater than first round');
    }
    return obj;
});
exports.AssetTransferTxnSchema = joi_1.default.object({
    tokenId: joi_1.default.number().required(),
    assetAmount: joi_1.default.custom(function (val) { return typeof val === 'number' || typeof val === 'bigint'; }).required(),
    receiver: addressSchema.required(),
});
exports.AssetToggleTxnSchema = joi_1.default.object({
    tokenId: joi_1.default.number().required(),
    receiver: addressSchema.required(),
});
