// bitgo/dist/src/v2/coins/fiat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fiat = void 0;
/**
 * @prettier
 */
const baseCoin_1 = require("../baseCoin");
const errors_1 = require("../../errors");
class Fiat extends baseCoin_1.BaseCoin {
    /**
     * Returns the factor between the base unit and its smallest subdivison
     * @return {number}
     */
    getBaseFactor() {
        return 1e2;
    }
    getChain() {
        return 'fiat';
    }
    getFamily() {
        return 'fiat';
    }
    getFullName() {
        return 'Fiat';
    }
    /**
     * Return whether the given m of n wallet signers/ key amounts are valid for the coin
     */
    isValidMofNSetup({ m, n }) {
        return m === 0 && n === 0;
    }
    isValidAddress(address) {
        throw new errors_1.MethodNotImplementedError();
    }
    generateKeyPair(seed) {
        throw new errors_1.MethodNotImplementedError();
    }
    isValidPub(pub) {
        throw new errors_1.MethodNotImplementedError();
    }
    async parseTransaction(params) {
        throw new errors_1.MethodNotImplementedError();
    }
    isWalletAddress(params) {
        throw new errors_1.MethodNotImplementedError();
    }
    async verifyTransaction(params) {
        throw new errors_1.MethodNotImplementedError();
    }
    async signTransaction(params = {}) {
        throw new errors_1.MethodNotImplementedError();
    }
}
exports.Fiat = Fiat;
