// bitgo/dist/src/v2/coins/ltc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ltc = void 0;
/**
 * @prettier
 */
const utxolib = require("@bitgo/utxo-lib");
const abstractUtxoCoin_1 = require("./abstractUtxoCoin");
const errors_1 = require("../../errors");
class Ltc extends abstractUtxoCoin_1.AbstractUtxoCoin {
    constructor(bitgo, network) {
        super(bitgo, network || utxolib.networks.litecoin);
        // use legacy script hash version, which is the current Bitcoin one
        this.altScriptHash = utxolib.networks.bitcoin.scriptHash;
        // do not support alt destinations in prod
        this.supportAltScriptDestination = false;
    }
    static createInstance(bitgo) {
        return new Ltc(bitgo);
    }
    getChain() {
        return 'ltc';
    }
    getFamily() {
        return 'ltc';
    }
    getFullName() {
        return 'Litecoin';
    }
    supportsBlockTarget() {
        return false;
    }
    /**
     * Canonicalize a Litecoin address for a specific scriptHash version
     * @param address
     * @param scriptHashVersion 1 or 2, where 1 is the old version and 2 is the new version
     * @returns {*} address string
     */
    canonicalAddress(address, scriptHashVersion = 2) {
        if (!this.isValidAddress(address, true)) {
            throw new errors_1.InvalidAddressError();
        }
        try {
            // try deserializing as bech32
            utxolib.address.fromBech32(address);
            // address may be all uppercase, but canonical bech32 addresses are all lowercase
            return address.toLowerCase();
        }
        catch (e) {
            // not a valid bech32, try to decode as base58
        }
        const addressDetails = utxolib.address.fromBase58Check(address, this.network);
        if (addressDetails.version === this.network.pubKeyHash) {
            // the pub keys never changed
            return address;
        }
        if ([1, 2].indexOf(scriptHashVersion) === -1) {
            throw new Error('scriptHashVersion needs to be either 1 or 2');
        }
        const scriptHashMap = {
            // altScriptHash is the old one
            1: this.altScriptHash,
            // by default we're using the new one
            2: this.network.scriptHash,
        };
        const newScriptHash = scriptHashMap[scriptHashVersion];
        return utxolib.address.toBase58Check(addressDetails.hash, newScriptHash, this.network);
    }
    calculateRecoveryAddress(scriptHashScript) {
        const bitgoAddress = utxolib.address.fromOutputScript(scriptHashScript, this.network);
        return this.canonicalAddress(bitgoAddress, 1);
    }
}
exports.Ltc = Ltc;
