// bitgo/dist/src/v2/coins/ofc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ofc = void 0;
/**
 * @prettier
 */
const crypto_1 = require("crypto");
const bip32 = require("bip32");
const errors_1 = require("../../errors");
const baseCoin_1 = require("../baseCoin");
class Ofc extends baseCoin_1.BaseCoin {
    static createInstance(bitgo) {
        return new Ofc(bitgo);
    }
    getChain() {
        return 'ofc';
    }
    /**
     * Generate secp256k1 key pair
     *
     * @param seed
     * @returns {Object} object with generated pub and prv
     */
    generateKeyPair(seed) {
        if (!seed) {
            // An extended private key has both a normal 256 bit private key and a 256
            // bit chain code, both of which must be random. 512 bits is therefore the
            // maximum entropy and gives us maximum security against cracking.
            seed = crypto_1.randomBytes(512 / 8);
        }
        const extendedKey = bip32.fromSeed(seed);
        const xpub = extendedKey.neutered().toBase58();
        return {
            pub: xpub,
            prv: extendedKey.toBase58(),
        };
    }
    getFamily() {
        return 'ofc';
    }
    getFullName() {
        return 'Offchain';
    }
    /**
     * Return whether the given m of n wallet signers/ key amounts are valid for the coin
     */
    isValidMofNSetup({ m, n }) {
        return m === 1 && n === 1;
    }
    /**
     * Return boolean indicating whether input is valid public key for the coin.
     *
     * @param {String} pub the pub to be checked
     * @returns {Boolean} is it valid?
     */
    isValidPub(pub) {
        try {
            return bip32.fromBase58(pub).isNeutered();
        }
        catch (e) {
            return false;
        }
    }
    isValidAddress(address) {
        throw new errors_1.MethodNotImplementedError();
    }
    getBaseFactor() {
        return 0;
    }
    async parseTransaction(params) {
        return {};
    }
    isWalletAddress(params) {
        throw new errors_1.MethodNotImplementedError();
    }
    async verifyTransaction(params) {
        return true;
    }
    async signTransaction(params) {
        throw new errors_1.MethodNotImplementedError();
    }
}
exports.Ofc = Ofc;
