// bitgo/dist/src/v2/coins/stellarToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StellarToken = void 0;
/**
 * @prettier
 */
const _ = require("lodash");
const xlm_1 = require("./xlm");
const sdk_core_1 = require("@bitgo/sdk-core");
const stellar = require("stellar-sdk");
class StellarToken extends xlm_1.Xlm {
    constructor(bitgo, tokenConfig) {
        super(bitgo);
        this.tokenConfig = tokenConfig;
        const [tokenCoin, token] = _.split(this.tokenConfig.type, xlm_1.Xlm.coinTokenPatternSeparator);
        if (tokenCoin !== tokenConfig.coin) {
            throw new sdk_core_1.BitGoJsError(`invalid coin found in token: ${this.tokenConfig.type}`);
        }
        if (!token || !token.match(StellarToken.tokenPattern)) {
            throw new sdk_core_1.BitGoJsError(`invalid token: ${this.tokenConfig.type}`);
        }
        [this._code, this._issuer] = _.split(token, '-');
    }
    static createTokenConstructor(config) {
        return (bitgo) => new StellarToken(bitgo, config);
    }
    get type() {
        return this.tokenConfig.type;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get code() {
        return this._code;
    }
    get issuer() {
        return this._issuer;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getStellarNetwork() {
        return this.tokenConfig.network === 'Testnet' ? stellar.Networks.TESTNET : stellar.Networks.PUBLIC;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getFullName() {
        return 'Stellar Token';
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return true;
    }
}
exports.StellarToken = StellarToken;
StellarToken.tokenPattern = /[A-Z]{1,12}-G[A-Z0-9]{55}/;
