// bitgo/dist/src/v2/coins/susd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Susd = void 0;
const baseCoin_1 = require("../baseCoin");
const errors_1 = require("../../errors");
class Susd extends baseCoin_1.BaseCoin {
    static createInstance(bitgo) {
        return new Susd(bitgo);
    }
    /**
     * Returns the factor between the base unit and its smallest subdivison
     * @return {number}
     */
    getBaseFactor() {
        return 1e2;
    }
    getChain() {
        return 'susd';
    }
    getFamily() {
        return 'susd';
    }
    getFullName() {
        return 'Silvergate USD';
    }
    /**
     * Return whether the given m of n wallet signers/ key amounts are valid for the coin
     */
    isValidMofNSetup({ m, n }) {
        return m === 0 && n === 0;
    }
    isValidAddress(address) {
        throw new errors_1.MethodNotImplementedError();
    }
    generateKeyPair(seed) {
        throw new errors_1.MethodNotImplementedError();
    }
    isValidPub(pub) {
        throw new errors_1.MethodNotImplementedError();
    }
    async parseTransaction(params) {
        return {};
    }
    isWalletAddress(params) {
        throw new errors_1.MethodNotImplementedError();
    }
    async verifyTransaction(params) {
        return true;
    }
    async signTransaction(params = {}) {
        throw new errors_1.MethodNotImplementedError();
    }
}
exports.Susd = Susd;
