// @bitgo/account-lib/dist/src/coin/cspr/walletInitializationBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletInitializationBuilder = void 0;
var casper_js_sdk_1 = require("casper-js-sdk");
var errors_1 = require("../baseCoin/errors");
var baseCoin_1 = require("../baseCoin");
var transactionBuilder_1 = require("./transactionBuilder");
var utils_1 = require("./utils");
var constants_1 = require("./constants");
var keyPair_1 = require("./keyPair");
var DEFAULT_OWNER_WEIGHT = 1;
var WalletInitializationBuilder = /** @class */ (function (_super) {
    __extends(WalletInitializationBuilder, _super);
    function WalletInitializationBuilder(_coinConfig) {
        var _this = _super.call(this, _coinConfig) || this;
        _this._owners = [];
        _this._contract = Uint8Array.from(Buffer.from(utils_1.casperContractHexCode, 'hex'));
        return _this;
    }
    // region Base Builder
    /** @inheritdoc */
    WalletInitializationBuilder.prototype.buildImplementation = function () {
        return __awaiter(this, void 0, void 0, function () {
            var args, extraArguments, index, ownerPublicKey, ownerAddress;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        args = this.buildWalletParameters();
                        extraArguments = new Map();
                        extraArguments.set(constants_1.TRANSACTION_TYPE, casper_js_sdk_1.CLValueBuilder.string(baseCoin_1.TransactionType[baseCoin_1.TransactionType.WalletInitialization]));
                        for (index = 0; index < this._owners.length; index++) {
                            ownerPublicKey = Buffer.from(this._owners[index].address.value()).toString('hex');
                            ownerAddress = new keyPair_1.KeyPair({ pub: ownerPublicKey }).getAddress();
                            extraArguments.set(constants_1.OWNER_PREFIX + index, casper_js_sdk_1.CLValueBuilder.string(ownerAddress));
                        }
                        this._session = { moduleBytes: this._contract, args: casper_js_sdk_1.RuntimeArgs.fromMap(args), extraArguments: extraArguments };
                        this.transaction.setTransactionType(baseCoin_1.TransactionType.WalletInitialization);
                        return [4 /*yield*/, _super.prototype.buildImplementation.call(this)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Build args needed to create a session, then we can send this session with the contract
     *
     * @returns {WalletInitContractArgs} contracts args to create a session
     */
    WalletInitializationBuilder.prototype.buildWalletParameters = function () {
        var accounts = this._owners.map(function (owner) { return casper_js_sdk_1.CLValueBuilder.byteArray(owner.address.toAccountHash()); });
        var weights = this._owners.map(function (owner) { return casper_js_sdk_1.CLValueBuilder.u8(owner.weight); });
        // set source address weight to zero to disable the master private key from signing.
        accounts.push(casper_js_sdk_1.CLValueBuilder.byteArray(casper_js_sdk_1.CLPublicKey.fromHex(this._source.address).toAccountHash()));
        weights.push(casper_js_sdk_1.CLValueBuilder.u8(0));
        return {
            action: casper_js_sdk_1.CLValueBuilder.string(constants_1.WALLET_INITIALIZATION_CONTRACT_ACTION),
            // This typo is on purpose since the contract we use for multisig wallet initialization expect this argument to be written like this.
            deployment_thereshold: casper_js_sdk_1.CLValueBuilder.u8(transactionBuilder_1.DEFAULT_N),
            key_management_threshold: casper_js_sdk_1.CLValueBuilder.u8(transactionBuilder_1.DEFAULT_M),
            accounts: casper_js_sdk_1.CLValueBuilder.list(accounts),
            weights: casper_js_sdk_1.CLValueBuilder.list(weights),
        };
    };
    /** @inheritdoc */
    WalletInitializationBuilder.prototype.initBuilder = function (tx) {
        _super.prototype.initBuilder.call(this, tx);
        this.transaction.setTransactionType(baseCoin_1.TransactionType.WalletInitialization);
        for (var ownerIndex = 0; ownerIndex < transactionBuilder_1.DEFAULT_M; ownerIndex++) {
            var ownerCLValue = tx.casperTx.session.getArgByName(constants_1.OWNER_PREFIX + ownerIndex);
            this.owner(ownerCLValue.value());
        }
    };
    // endregion
    // region Common builder methods
    /**
     * Set one of the owners of the multisig wallet.
     *
     * @param {string} address The public key of the owner's account
     * @returns {WalletInitializationBuilder} This wallet initialization builder
     */
    WalletInitializationBuilder.prototype.owner = function (address) {
        if (this._owners.length >= transactionBuilder_1.DEFAULT_M) {
            throw new errors_1.BuildTransactionError('A maximum of ' + transactionBuilder_1.DEFAULT_M + ' owners can be set for a multisig wallet');
        }
        this.validateAddress({ address: address });
        for (var _i = 0, _a = this._owners; _i < _a.length; _i++) {
            var _owner = _a[_i];
            if (address.substr(0, 2) + Buffer.from(_owner.address.value()).toString('hex') === address) {
                throw new errors_1.BuildTransactionError('Duplicated owner: ' + address);
            }
        }
        this._owners.push({ address: casper_js_sdk_1.CLPublicKey.fromHex(address), weight: DEFAULT_OWNER_WEIGHT });
        return this;
    };
    // endregion
    // region Validators
    WalletInitializationBuilder.prototype.validateMandatoryFields = function () {
        if (this._owners.length === 0) {
            throw new errors_1.BuildTransactionError('Invalid transaction: missing wallet owners');
        }
        if (this._owners.length !== transactionBuilder_1.DEFAULT_M) {
            throw new errors_1.BuildTransactionError("Invalid transaction: wrong number of owners -- required: " + transactionBuilder_1.DEFAULT_M + ", found: " + this._owners.length);
        }
        _super.prototype.validateMandatoryFields.call(this);
    };
    return WalletInitializationBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.WalletInitializationBuilder = WalletInitializationBuilder;
