// @bitgo/account-lib/dist/src/coin/celo/stakingUtils.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOperationConfig = exports.WithdrawMethodId = exports.ActivateMethodId = exports.UnvoteMethodId = exports.VoteMethodId = exports.UnlockMethodId = exports.LockMethodId = void 0;
var statics_1 = require("@bitgo/statics");
var baseCoin_1 = require("../baseCoin");
exports.LockMethodId = '0xf83d08ba'; // lock()
exports.UnlockMethodId = '0x6198e339'; // unlock()
exports.VoteMethodId = '0x580d747a'; // vote()
exports.UnvoteMethodId = '0x6e198475'; // revokeActive()
exports.ActivateMethodId = '0x1c5a9d9c'; // activate()
exports.WithdrawMethodId = '0x2e1a7d4d'; // withdraw()
var LockedGoldTestnetAddress = '0x94c3e6675015d8479b648657e7ddfcd938489d0d';
var LockedGoldMainnetAddress = '0x6cc083aed9e3ebe302a6336dbc7c921c9f03349e';
var ElectionTestnetAddress = '0x11fe523f93cac185d12cb39cc3bd279d2de524f8';
var ElectionMainnetAddress = '0x8d6677192144292870907e3fa8a5527fe55a7ff6';
var operations = (_a = {},
    _a[baseCoin_1.StakingOperationTypes.LOCK] = (_b = {},
        _b[statics_1.NetworkType.TESTNET] = {
            contractAddress: LockedGoldTestnetAddress,
            methodId: exports.LockMethodId,
            types: [],
        },
        _b[statics_1.NetworkType.MAINNET] = {
            contractAddress: LockedGoldMainnetAddress,
            methodId: exports.LockMethodId,
            types: [],
        },
        _b),
    _a[baseCoin_1.StakingOperationTypes.UNLOCK] = (_c = {},
        _c[statics_1.NetworkType.TESTNET] = {
            contractAddress: LockedGoldTestnetAddress,
            methodId: exports.UnlockMethodId,
            types: ['uint256'],
        },
        _c[statics_1.NetworkType.MAINNET] = {
            contractAddress: LockedGoldMainnetAddress,
            methodId: exports.UnlockMethodId,
            types: ['uint256'],
        },
        _c),
    _a[baseCoin_1.StakingOperationTypes.VOTE] = (_d = {},
        _d[statics_1.NetworkType.TESTNET] = {
            contractAddress: ElectionTestnetAddress,
            methodId: exports.VoteMethodId,
            types: ['address', 'uint256', 'address', 'address'],
        },
        _d[statics_1.NetworkType.MAINNET] = {
            contractAddress: ElectionMainnetAddress,
            methodId: exports.VoteMethodId,
            types: ['address', 'uint256', 'address', 'address'],
        },
        _d),
    _a[baseCoin_1.StakingOperationTypes.UNVOTE] = (_e = {},
        _e[statics_1.NetworkType.TESTNET] = {
            contractAddress: ElectionTestnetAddress,
            methodId: exports.UnvoteMethodId,
            types: ['address', 'uint256', 'address', 'address', 'uint256'],
        },
        _e[statics_1.NetworkType.MAINNET] = {
            contractAddress: ElectionMainnetAddress,
            methodId: exports.UnvoteMethodId,
            types: ['address', 'uint256', 'address', 'address', 'uint256'],
        },
        _e),
    _a[baseCoin_1.StakingOperationTypes.ACTIVATE] = (_f = {},
        _f[statics_1.NetworkType.TESTNET] = {
            contractAddress: ElectionTestnetAddress,
            methodId: exports.ActivateMethodId,
            types: ['address'],
        },
        _f[statics_1.NetworkType.MAINNET] = {
            contractAddress: ElectionMainnetAddress,
            methodId: exports.ActivateMethodId,
            types: ['address'],
        },
        _f),
    _a[baseCoin_1.StakingOperationTypes.WITHDRAW] = (_g = {},
        _g[statics_1.NetworkType.TESTNET] = {
            contractAddress: LockedGoldTestnetAddress,
            methodId: exports.WithdrawMethodId,
            types: ['uint256'],
        },
        _g[statics_1.NetworkType.MAINNET] = {
            contractAddress: LockedGoldMainnetAddress,
            methodId: exports.WithdrawMethodId,
            types: ['uint256'],
        },
        _g),
    _a);
/**
 * @param type
 * @param network
 */
function getOperationConfig(type, network) {
    return operations[type][network];
}
exports.getOperationConfig = getOperationConfig;
