// @bitgo/account-lib/dist/src/coin/near/contractCallWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractCallWrapper = void 0;
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var errors_1 = require("../baseCoin/errors");
/**
 * Contains parameters to call a Near Function Call action
 */
var ContractCallWrapper = /** @class */ (function () {
    function ContractCallWrapper() {
    }
    Object.defineProperty(ContractCallWrapper.prototype, "methodName", {
        /** Get method contract name */
        get: function () {
            return this._methodName;
        },
        /** Set method contract name */
        set: function (methodName) {
            this._methodName = methodName;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ContractCallWrapper.prototype, "gas", {
        /** Get gas, expresed on yocto*/
        get: function () {
            return this._gas;
        },
        /** Set gas, expresed on yocto */
        set: function (gas) {
            if (!this.isValidAmount(new bignumber_js_1.default(gas))) {
                throw new errors_1.InvalidParameterValueError('Invalid gas value');
            }
            this._gas = gas;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ContractCallWrapper.prototype, "deposit", {
        /** Get deposit, expresed on yocto */
        get: function () {
            return this._deposit;
        },
        /** Set deposit, expresed on yocto */
        set: function (deposit) {
            if (!this.isValidAmount(new bignumber_js_1.default(deposit))) {
                throw new errors_1.InvalidParameterValueError('Invalid deposit value');
            }
            this._deposit = deposit;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ContractCallWrapper.prototype, "args", {
        /** Set args, which are the parameters of a method */
        get: function () {
            return this._args;
        },
        /** Get args, which are the parameters of a method */
        set: function (args) {
            this._args = args;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Get all parameters of the contractCallWrapper
     * @returns {FunctionCall}
     */
    ContractCallWrapper.prototype.getParams = function () {
        return {
            methodName: this._methodName,
            args: this._args,
            gas: this._gas,
            deposit: this._deposit,
        };
    };
    ContractCallWrapper.prototype.isValidAmount = function (value) {
        return !value.isLessThan(0);
    };
    return ContractCallWrapper;
}());
exports.ContractCallWrapper = ContractCallWrapper;
