// @bitgo/account-lib/dist/src/coin/near/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
var keyPair_1 = require("./keyPair");
var crypto_1 = require("./../../utils/crypto");
var bs58_1 = __importDefault(require("bs58"));
var Utils = /** @class */ (function () {
    function Utils() {
    }
    /** @inheritdoc */
    Utils.prototype.isValidAddress = function (address) {
        return this.isValidAccountId(address);
    };
    /** @inheritdoc */
    Utils.prototype.isValidBlockId = function (hash) {
        return crypto_1.isBase58(hash, 32);
    };
    /** @inheritdoc */
    Utils.prototype.isValidPrivateKey = function (key) {
        return this.isValidKey(key);
    };
    /** @inheritdoc */
    Utils.prototype.isValidPublicKey = function (pubKey) {
        return this.isValidAccountId(pubKey);
    };
    /**
     * Check if the key have the correct format
     *
     * @param {string} key - string to be checked
     * @return {boolean} true if the string have the correct format otherwise return false
     */
    Utils.prototype.isValidKey = function (key) {
        try {
            new keyPair_1.KeyPair({ prv: key });
            return true;
        }
        catch (_a) {
            return false;
        }
    };
    /**
     * Check if the account have the correct format
     *
     * @param {string} accountId - string to be checked
     * @return {boolean} if the string can have the correct format and match the expected length
     * or can create a new keyPair
     */
    Utils.prototype.isValidAccountId = function (accountId) {
        try {
            // this regex is from near doc, https://docs.near.org/docs/concepts/account
            (/^(([a-z\d]+[\-_])*[a-z\d]+\.)*([a-z\d]+[\-_])*[a-z\d]+$/.test(accountId) &&
                accountId.length > 2 &&
                accountId.length < 64) ||
                new keyPair_1.KeyPair({ pub: accountId });
            return true;
        }
        catch (_a) {
            return false;
        }
    };
    /** @inheritdoc */
    Utils.prototype.isValidSignature = function (signature) {
        return crypto_1.isBase58(signature, 64);
    };
    /** @inheritdoc */
    Utils.prototype.isValidTransactionId = function (txId) {
        return crypto_1.isBase58(txId, 32);
    };
    Utils.prototype.base58Encode = function (value) {
        return bs58_1.default.encode(value);
    };
    /**
     * Check if base58 decoded string is equal to length
     *
     * @param {string} value - string to be checked
     * @param {number} length - expected decoded length
     * @return {boolean} if the string can decoded as base58 and match the expected length
     */
    Utils.prototype.isBase58 = function (value, length) {
        try {
            return !!value && bs58_1.default.decode(value).length === length;
        }
        catch (e) {
            return false;
        }
    };
    return Utils;
}());
exports.Utils = Utils;
var utils = new Utils();
exports.default = utils;
