// @bitgo/account-lib/dist/src/coin/xrp/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
var errors_1 = require("../baseCoin/errors");
var Utils = /** @class */ (function () {
    function Utils() {
    }
    /** @inheritdoc */
    Utils.prototype.isValidAddress = function (address) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    /** @inheritdoc */
    Utils.prototype.isValidBlockId = function (hash) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    /** @inheritdoc */
    Utils.prototype.isValidPrivateKey = function (key) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    /** @inheritdoc */
    Utils.prototype.isValidPublicKey = function (key) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    /** @inheritdoc */
    Utils.prototype.isValidSignature = function (signature) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    /** @inheritdoc */
    Utils.prototype.isValidTransactionId = function (txId) {
        throw new errors_1.NotImplementedError('method not implemented');
    };
    return Utils;
}());
exports.Utils = Utils;
