// @bitgo/account-lib/dist/src/coin/sol/ataInitializationTransaction.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtaInitializationTransaction = void 0;
var _1 = require(".");
var instructionParamsFactory_1 = require("./instructionParamsFactory");
var constants_1 = require("./constants");
var utils_1 = require("./utils");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var errors_1 = require("../baseCoin/errors");
var TRANSFER_AMOUNT_UNKNOWN_TEXT = 'TRANSFER_AMOUNT_UNKNOWN';
var AtaInitializationTransaction = /** @class */ (function (_super) {
    __extends(AtaInitializationTransaction, _super);
    function AtaInitializationTransaction(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    Object.defineProperty(AtaInitializationTransaction.prototype, "tokenAccountRentExemptAmount", {
        get: function () {
            return this._tokenAccountRentExemptAmount;
        },
        set: function (amount) {
            this._tokenAccountRentExemptAmount = amount;
        },
        enumerable: false,
        configurable: true
    });
    /** @inheritDoc */
    AtaInitializationTransaction.prototype.loadInputsAndOutputs = function () {
        var _a;
        if (!this._solTransaction || ((_a = this._solTransaction.instructions) === null || _a === void 0 ? void 0 : _a.length) === 0) {
            return;
        }
        var outputs = [];
        var inputs = [];
        var instructionParams = instructionParamsFactory_1.instructionParamsFactory(this.type, this._solTransaction.instructions);
        for (var _i = 0, instructionParams_1 = instructionParams; _i < instructionParams_1.length; _i++) {
            var instruction = instructionParams_1[_i];
            if (instruction.type === constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount) {
                var token = utils_1.getSolTokenFromAddress(instruction.params.mintAddress, this._coinConfig.network);
                if (!token) {
                    throw new errors_1.NotSupported('Invalid transaction, token mint address not supported: ' + instruction.params.mintAddress);
                }
                inputs.push({
                    address: instruction.params.ownerAddress,
                    value: this.tokenAccountRentExemptAmount || TRANSFER_AMOUNT_UNKNOWN_TEXT,
                    coin: token.name,
                });
                outputs.push({
                    address: instruction.params.ataAddress,
                    value: this.tokenAccountRentExemptAmount || TRANSFER_AMOUNT_UNKNOWN_TEXT,
                    coin: token.name,
                });
            }
        }
        this._outputs = outputs;
        this._inputs = inputs;
    };
    /** @inheritDoc */
    AtaInitializationTransaction.prototype.explainTransaction = function () {
        var decodedInstructions = instructionParamsFactory_1.instructionParamsFactory(this._type, this._solTransaction.instructions);
        var memo = undefined;
        var outputAmount = new bignumber_js_1.default(0);
        var outputs = [];
        for (var _i = 0, decodedInstructions_1 = decodedInstructions; _i < decodedInstructions_1.length; _i++) {
            var instruction = decodedInstructions_1[_i];
            switch (instruction.type) {
                case constants_1.InstructionBuilderTypes.Memo:
                    memo = instruction.params.memo;
                    break;
                case constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount:
                    var createAtaInstruction = instruction;
                    outputs.push({
                        address: createAtaInstruction.params.ataAddress,
                        amount: this.tokenAccountRentExemptAmount || TRANSFER_AMOUNT_UNKNOWN_TEXT,
                    });
                    if (this.tokenAccountRentExemptAmount) {
                        outputAmount = outputAmount.plus(this.tokenAccountRentExemptAmount);
                    }
                    break;
            }
        }
        return this.getExplainedTransaction(outputAmount, outputs, memo);
    };
    return AtaInitializationTransaction;
}(_1.Transaction));
exports.AtaInitializationTransaction = AtaInitializationTransaction;
