// @bitgo/account-lib/dist/src/coin/sol/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ataInitInstructionIndexes = exports.stakingWithdrawInstructionsIndexes = exports.stakingDeactivateInstructionsIndexes = exports.stakingActivateInstructionsIndexes = exports.walletInitInstructionIndexes = exports.VALID_SYSTEM_INSTRUCTION_TYPES = exports.InstructionBuilderTypes = exports.ValidInstructionTypesEnum = exports.MAX_MEMO_LENGTH = exports.SEED_LENGTH = exports.MEMO_PROGRAM_PK = void 0;
exports.MEMO_PROGRAM_PK = 'MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr';
exports.SEED_LENGTH = 32;
exports.MAX_MEMO_LENGTH = 100;
// Sdk instructions, mainly to check decoded types.
var ValidInstructionTypesEnum;
(function (ValidInstructionTypesEnum) {
    ValidInstructionTypesEnum["AdvanceNonceAccount"] = "AdvanceNonceAccount";
    ValidInstructionTypesEnum["Create"] = "Create";
    ValidInstructionTypesEnum["InitializeNonceAccount"] = "InitializeNonceAccount";
    ValidInstructionTypesEnum["StakingInitialize"] = "Initialize";
    ValidInstructionTypesEnum["StakingDelegate"] = "Delegate";
    ValidInstructionTypesEnum["StakingDeactivate"] = "Deactivate";
    ValidInstructionTypesEnum["StakingWithdraw"] = "Withdraw";
    ValidInstructionTypesEnum["Transfer"] = "Transfer";
    ValidInstructionTypesEnum["TokenTransfer"] = "TokenTransfer";
    ValidInstructionTypesEnum["Memo"] = "Memo";
    ValidInstructionTypesEnum["InitializeAssociatedTokenAccount"] = "InitializeAssociatedTokenAccount";
})(ValidInstructionTypesEnum = exports.ValidInstructionTypesEnum || (exports.ValidInstructionTypesEnum = {}));
// Internal instructions types
var InstructionBuilderTypes;
(function (InstructionBuilderTypes) {
    InstructionBuilderTypes["CreateNonceAccount"] = "CreateNonceAccount";
    InstructionBuilderTypes["StakingActivate"] = "Activate";
    InstructionBuilderTypes["StakingDeactivate"] = "Deactivate";
    InstructionBuilderTypes["StakingWithdraw"] = "Withdraw";
    InstructionBuilderTypes["Transfer"] = "Transfer";
    InstructionBuilderTypes["Memo"] = "Memo";
    InstructionBuilderTypes["NonceAdvance"] = "NonceAdvance";
    InstructionBuilderTypes["CreateAssociatedTokenAccount"] = "CreateAssociatedTokenAccount";
    InstructionBuilderTypes["TokenTransfer"] = "TokenTransfer";
})(InstructionBuilderTypes = exports.InstructionBuilderTypes || (exports.InstructionBuilderTypes = {}));
exports.VALID_SYSTEM_INSTRUCTION_TYPES = [
    ValidInstructionTypesEnum.AdvanceNonceAccount,
    ValidInstructionTypesEnum.Create,
    ValidInstructionTypesEnum.StakingInitialize,
    ValidInstructionTypesEnum.StakingDelegate,
    ValidInstructionTypesEnum.StakingDeactivate,
    ValidInstructionTypesEnum.StakingWithdraw,
    ValidInstructionTypesEnum.Transfer,
    ValidInstructionTypesEnum.InitializeNonceAccount,
    ValidInstructionTypesEnum.Memo,
    ValidInstructionTypesEnum.InitializeAssociatedTokenAccount,
    ValidInstructionTypesEnum.TokenTransfer,
];
/** Const to check the order of the Wallet Init instructions when decode */
exports.walletInitInstructionIndexes = {
    Create: 0,
    InitializeNonceAccount: 1,
    Memo: 2,
};
/** Const to check the order of the Staking Activate instructions when decode */
exports.stakingActivateInstructionsIndexes = {
    Create: 0,
    Initialize: 1,
    Delegate: 2,
    Memo: 3,
};
/** Const to check the order of the Staking Deactivate instructions when decode */
exports.stakingDeactivateInstructionsIndexes = {
    Deactivate: 0,
    Memo: 1,
};
/** Const to check the order of the Staking Withdraw instructions when decode */
exports.stakingWithdrawInstructionsIndexes = {
    Withdraw: 0,
    Memo: 1,
};
/** Const to check the order of the ATA init instructions when decode */
exports.ataInitInstructionIndexes = {
    InitializeAssociatedTokenAccount: 0,
    Memo: 1,
};
