// @bitgo/account-lib/dist/src/coin/sol/keyPair.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
var web3_js_1 = require("@solana/web3.js");
var baseCoin_1 = require("../baseCoin");
var utils_1 = require("./utils");
var KeyPair = /** @class */ (function (_super) {
    __extends(KeyPair, _super);
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    function KeyPair(source) {
        return _super.call(this, source) || this;
    }
    /** @inheritdoc */
    KeyPair.prototype.recordKeysFromPrivateKeyInProtocolFormat = function (prv) {
        var prvKey = utils_1.base58ToUint8Array(prv);
        var keyPair = web3_js_1.Keypair.fromSecretKey(prvKey);
        return {
            pub: keyPair.publicKey.toBuffer().toString('hex'),
            prv: Buffer.from(keyPair.secretKey.slice(0, 32)).toString('hex'),
        };
    };
    /** @inheritdoc */
    KeyPair.prototype.recordKeysFromPublicKeyInProtocolFormat = function (pub) {
        return {
            pub: Buffer.from(utils_1.base58ToUint8Array(pub)).toString('hex'),
        };
    };
    /**
     * Solana default keys format public key as a base58 string and secret key as Uint8Array
     *
     * @param {boolean} raw defines if the prv key is returned in Uint8Array, default is base58
     * @returns {SolanaKeys} The keys in the defined format
     */
    KeyPair.prototype.getKeys = function (raw) {
        if (raw === void 0) { raw = false; }
        // keys are originally created in hex, but we need base58
        var publicKeyBuffer = Buffer.from(this.keyPair.pub, 'hex');
        var base58Pub = utils_1.Uint8ArrayTobase58(publicKeyBuffer);
        var result = { pub: base58Pub };
        if (!!this.keyPair.prv) {
            var secretKeyBuffer = Buffer.from(this.keyPair.prv, 'hex');
            var solanaSecretKey = new Uint8Array(64);
            solanaSecretKey.set(secretKeyBuffer);
            solanaSecretKey.set(publicKeyBuffer, 32);
            if (raw) {
                result.prv = solanaSecretKey;
            }
            else {
                result.prv = utils_1.Uint8ArrayTobase58(solanaSecretKey);
            }
        }
        return result;
    };
    /** @inheritdoc */
    KeyPair.prototype.getAddress = function () {
        var keys = this.getKeys();
        return keys.pub;
    };
    return KeyPair;
}(baseCoin_1.Ed25519KeyPair));
exports.KeyPair = KeyPair;
