// @bitgo/account-lib/dist/src/coin/sol/solInstructionFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.solInstructionFactory = void 0;
var web3_js_1 = require("@solana/web3.js");
var assert_1 = __importDefault(require("assert"));
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var constants_1 = require("./constants");
var statics_1 = require("@bitgo/statics");
var spl_token_1 = require("@solana/spl-token");
/**
 * Construct Solana instructions from instructions params
 *
 * @param {InstructionParams} instructionToBuild - the data containing the instruction params
 * @returns {TransactionInstruction[]} An array containing supported Solana instructions
 */
function solInstructionFactory(instructionToBuild) {
    switch (instructionToBuild.type) {
        case constants_1.InstructionBuilderTypes.NonceAdvance:
            return advanceNonceInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.Memo:
            return memoInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.Transfer:
            return transferInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.TokenTransfer:
            return tokenTransferInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.CreateNonceAccount:
            return createNonceAccountInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingActivate:
            return stakingInitializeInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingDeactivate:
            return stakingDeactivateInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.StakingWithdraw:
            return stakingWithdrawInstruction(instructionToBuild);
        case constants_1.InstructionBuilderTypes.CreateAssociatedTokenAccount:
            return createATAInstruction(instructionToBuild);
        default:
            throw new Error("Invalid instruction type or not supported");
    }
}
exports.solInstructionFactory = solInstructionFactory;
/**
 * Construct Advance Nonce Solana instructions
 *
 * @param {Nonce} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Advance Nonce Solana instruction
 */
function advanceNonceInstruction(data) {
    var _a = data.params, authWalletAddress = _a.authWalletAddress, walletNonceAddress = _a.walletNonceAddress;
    assert_1.default(authWalletAddress, 'Missing authWalletAddress param');
    assert_1.default(walletNonceAddress, 'Missing walletNonceAddress param');
    var nonceInstruction = web3_js_1.SystemProgram.nonceAdvance({
        noncePubkey: new web3_js_1.PublicKey(walletNonceAddress),
        authorizedPubkey: new web3_js_1.PublicKey(authWalletAddress),
    });
    return [nonceInstruction];
}
/**
 * Construct Memo Solana instructions
 *
 * @param {Memo} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Memo Solana instruction
 */
function memoInstruction(data) {
    var memo = data.params.memo;
    assert_1.default(memo, 'Missing memo param');
    var memoInstruction = new web3_js_1.TransactionInstruction({
        keys: [],
        programId: new web3_js_1.PublicKey(constants_1.MEMO_PROGRAM_PK),
        data: Buffer.from(memo),
    });
    return [memoInstruction];
}
/**
 * Construct Transfer Solana instructions
 *
 * @param {Transfer} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Transfer Solana instruction
 */
function transferInstruction(data) {
    var _a = data.params, fromAddress = _a.fromAddress, toAddress = _a.toAddress, amount = _a.amount;
    assert_1.default(fromAddress, 'Missing fromAddress param');
    assert_1.default(toAddress, 'Missing toAddress param');
    assert_1.default(amount, 'Missing toAddress param');
    var transferInstruction = web3_js_1.SystemProgram.transfer({
        fromPubkey: new web3_js_1.PublicKey(fromAddress),
        toPubkey: new web3_js_1.PublicKey(toAddress),
        lamports: new bignumber_js_1.default(amount).toNumber(),
    });
    return [transferInstruction];
}
/**
 * Construct Transfer Solana instructions
 *
 * @param {Transfer} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Transfer Solana instruction
 */
function tokenTransferInstruction(data) {
    var _a = data.params, fromAddress = _a.fromAddress, toAddress = _a.toAddress, amount = _a.amount, tokenName = _a.tokenName, sourceAddress = _a.sourceAddress;
    assert_1.default(fromAddress, 'Missing fromAddress (owner) param');
    assert_1.default(toAddress, 'Missing toAddress param');
    assert_1.default(amount, 'Missing amount param');
    assert_1.default(tokenName, 'Missing token name');
    assert_1.default(sourceAddress, 'Missing ata address');
    var token = statics_1.coins.get(data.params.tokenName);
    assert_1.default(token instanceof statics_1.SolCoin);
    var transferInstruction = spl_token_1.Token.createTransferCheckedInstruction(spl_token_1.TOKEN_PROGRAM_ID, new web3_js_1.PublicKey(sourceAddress), new web3_js_1.PublicKey(token.tokenAddress), new web3_js_1.PublicKey(toAddress), new web3_js_1.PublicKey(fromAddress), [], new bignumber_js_1.default(amount).toNumber(), token.decimalPlaces);
    return [transferInstruction];
}
/**
 * Construct Create and Initialize Nonce Solana instructions
 *
 * @param {WalletInit} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Create and Initialize Nonce Solana instruction
 */
function createNonceAccountInstruction(data) {
    var _a = data.params, fromAddress = _a.fromAddress, nonceAddress = _a.nonceAddress, authAddress = _a.authAddress, amount = _a.amount;
    assert_1.default(fromAddress, 'Missing fromAddress param');
    assert_1.default(nonceAddress, 'Missing nonceAddress param');
    assert_1.default(authAddress, 'Missing authAddress param');
    assert_1.default(amount, 'Missing amount param');
    var nonceAccountInstruction = web3_js_1.SystemProgram.createNonceAccount({
        fromPubkey: new web3_js_1.PublicKey(fromAddress),
        noncePubkey: new web3_js_1.PublicKey(nonceAddress),
        authorizedPubkey: new web3_js_1.PublicKey(authAddress),
        lamports: new bignumber_js_1.default(amount).toNumber(),
    });
    return nonceAccountInstruction.instructions;
}
/**
 * Construct Create Staking Account and Delegate Solana instructions
 *
 * @param {StakingActivate} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Create Staking Account and Delegate Solana instructions
 */
function stakingInitializeInstruction(data) {
    var _a = data.params, fromAddress = _a.fromAddress, stakingAddress = _a.stakingAddress, amount = _a.amount, validator = _a.validator;
    assert_1.default(fromAddress, 'Missing fromAddress param');
    assert_1.default(stakingAddress, 'Missing stakingAddress param');
    assert_1.default(amount, 'Missing amount param');
    assert_1.default(validator, 'Missing validator param');
    var fromPubkey = new web3_js_1.PublicKey(fromAddress);
    var stakePubkey = new web3_js_1.PublicKey(stakingAddress);
    var tx = new web3_js_1.Transaction();
    var walletInitStaking = web3_js_1.StakeProgram.createAccount({
        fromPubkey: fromPubkey,
        stakePubkey: stakePubkey,
        authorized: new web3_js_1.Authorized(fromPubkey, fromPubkey),
        lockup: new web3_js_1.Lockup(0, 0, fromPubkey),
        lamports: new bignumber_js_1.default(amount).toNumber(),
    });
    tx.add(walletInitStaking);
    var delegateStaking = web3_js_1.StakeProgram.delegate({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
        votePubkey: new web3_js_1.PublicKey(validator),
    });
    tx.add(delegateStaking);
    return tx.instructions;
}
/**
 * Construct staking deactivate Solana instructions
 *
 * @param {StakingDeactivate} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing staking deactivate instruction
 */
function stakingDeactivateInstruction(data) {
    var _a = data.params, fromAddress = _a.fromAddress, stakingAddress = _a.stakingAddress;
    assert_1.default(fromAddress, 'Missing fromAddress param');
    assert_1.default(stakingAddress, 'Missing stakingAddress param');
    var deactivateStaking = web3_js_1.StakeProgram.deactivate({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
    });
    return deactivateStaking.instructions;
}
/**
 * Construct Staking Withdraw Solana instructions
 *
 * @param {StakingWithdraw} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Staking Withdraw  Solana instructions
 */
function stakingWithdrawInstruction(data) {
    var _a = data.params, fromAddress = _a.fromAddress, stakingAddress = _a.stakingAddress, amount = _a.amount;
    assert_1.default(fromAddress, 'Missing fromAddress param');
    assert_1.default(stakingAddress, 'Missing stakingAddress param');
    assert_1.default(amount, 'Missing amount param');
    var withdrawStaking = web3_js_1.StakeProgram.withdraw({
        stakePubkey: new web3_js_1.PublicKey(stakingAddress),
        authorizedPubkey: new web3_js_1.PublicKey(fromAddress),
        toPubkey: new web3_js_1.PublicKey(fromAddress),
        lamports: new bignumber_js_1.default(amount).toNumber(),
    });
    return withdrawStaking.instructions;
}
/**
 * Construct Create and Initialize Nonce Solana instructions
 *
 * @param {WalletInit} data - the data to build the instruction
 * @returns {TransactionInstruction[]} An array containing Create and Initialize Nonce Solana instruction
 */
function createATAInstruction(data) {
    var _a = data.params, mintAddress = _a.mintAddress, ataAddress = _a.ataAddress, ownerAddress = _a.ownerAddress, payerAddress = _a.payerAddress;
    assert_1.default(mintAddress, 'Missing mintAddress param');
    assert_1.default(ataAddress, 'Missing ataAddress param');
    assert_1.default(ownerAddress, 'Missing ownerAddress param');
    assert_1.default(payerAddress, 'Missing payerAddress param');
    var associatedTokenAccountInstruction = spl_token_1.Token.createAssociatedTokenAccountInstruction(spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID, spl_token_1.TOKEN_PROGRAM_ID, new web3_js_1.PublicKey(mintAddress), new web3_js_1.PublicKey(ataAddress), new web3_js_1.PublicKey(ownerAddress), new web3_js_1.PublicKey(payerAddress));
    return [associatedTokenAccountInstruction];
}
