// bitgo/dist/src/legacyBitcoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.address = exports.networks = exports.hdPath = exports.HDNode = exports.makeRandomKey = void 0;
/**
 * This file contains a compatability layer for some deprecated types and helper methods
 *
 * @prettier
 */
const utxolib = require("@bitgo/utxo-lib");
const bip32 = require("bip32");
const sdk_api_1 = require("@bitgo/sdk-api");
const bitcoinUtil = require("./bitcoin");
function createECPairCompat(ecPair) {
    return Object.assign(ecPair, {
        getPublicKeyBuffer() {
            return ecPair.publicKey;
        },
    });
}
function makeRandomKey() {
    return createECPairCompat(bitcoinUtil.makeRandomKey());
}
exports.makeRandomKey = makeRandomKey;
/**
 * Implementation of legacy "HDNode" class as used by certain components
 */
class HDNode {
    constructor(bip32) {
        this.bip32 = bip32;
    }
    get publicKey() {
        return this.bip32.publicKey;
    }
    get privateKey() {
        return this.bip32.privateKey;
    }
    static fromSeedBuffer(buf) {
        return new HDNode(bip32.fromSeed(buf));
    }
    static fromBase58(str) {
        return new HDNode(bip32.fromBase58(str));
    }
    isNeutered() {
        return this.bip32.isNeutered();
    }
    neutered() {
        return new HDNode(this.bip32.neutered());
    }
    toBase58() {
        return this.bip32.toBase58();
    }
    derivePath(p) {
        return new HDNode(this.bip32.derivePath(sdk_api_1.sanitizeLegacyPath(p)));
    }
}
exports.HDNode = HDNode;
function hdPath(hdNode) {
    return {
        derive(path) {
            return hdNode.derivePath(path);
        },
        deriveKey(path) {
            const node = hdNode.derivePath(path);
            if (node.privateKey) {
                return createECPairCompat(utxolib.ECPair.fromPrivateKey(node.privateKey));
            }
            else {
                return createECPairCompat(utxolib.ECPair.fromPublicKey(node.publicKey));
            }
        },
    };
}
exports.hdPath = hdPath;
exports.networks = utxolib.networks;
exports.address = {
    fromBase58Check(addr) {
        return utxolib.address.fromBase58Check(addr, utxolib.networks.bitcoin);
    },
    toBase58Check(hash, version) {
        return utxolib.address.toBase58Check(hash, version, utxolib.networks.bitcoin);
    },
};
