// bitgo/dist/src/v2/coinFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalCoinFactory = exports.CoinFactory = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const coins_1 = require("./coins");
const config_1 = require("../config");
const errors = require("../errors");
const bcha_1 = require("./coins/bcha");
const tbcha_1 = require("./coins/tbcha");
const dot_1 = require("./coins/dot");
const tdot_1 = require("./coins/tdot");
const near_1 = require("./coins/near");
const tnear_1 = require("./coins/tnear");
class CoinFactory {
    constructor() {
        this.coinConstructors = new Map();
    }
    /**
     *
     * @param bitgo
     * @param name
     * @throws CoinNotDefinedError
     * @throws UnsupportedCoinError
     */
    getInstance(bitgo, name) {
        let staticsCoin;
        try {
            staticsCoin = statics_1.coins.get(name);
        }
        catch (e) {
            if (!(e instanceof statics_1.CoinNotDefinedError)) {
                throw e;
            }
        }
        const constructor = this.coinConstructors.get(name);
        if (constructor) {
            return constructor(bitgo, staticsCoin);
        }
        const ethConstructor = this.coinConstructors.get('eth');
        if (ethConstructor) {
            const ethCoin = ethConstructor(bitgo, staticsCoin);
            if (ethCoin.isValidAddress(name)) {
                const unknownTokenConstructor = coins_1.Erc20Token.createTokenConstructor({
                    type: 'unknown',
                    coin: 'eth',
                    network: 'Mainnet',
                    name: 'Unknown',
                    tokenContractAddress: name,
                    decimalPlaces: 0,
                });
                return unknownTokenConstructor(bitgo);
            }
        }
        throw new errors.UnsupportedCoinError(name);
    }
    registerCoinConstructor(name, constructor) {
        if (this.coinConstructors.has(name)) {
            throw new Error(`coin '${name}' is already defined`);
        }
        this.coinConstructors.set(name, constructor);
    }
}
exports.CoinFactory = CoinFactory;
exports.GlobalCoinFactory = new CoinFactory();
exports.GlobalCoinFactory.registerCoinConstructor('btc', coins_1.Btc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tbtc', coins_1.Tbtc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('bch', coins_1.Bch.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tbch', coins_1.Tbch.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('bcha', bcha_1.Bcha.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tbcha', tbcha_1.Tbcha.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('bsv', coins_1.Bsv.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tbsv', coins_1.Tbsv.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('btg', coins_1.Btg.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('dot', dot_1.Dot.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tdot', tdot_1.Tdot.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('ltc', coins_1.Ltc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tltc', coins_1.Tltc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('eos', coins_1.Eos.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('teos', coins_1.Teos.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('eth', coins_1.Eth.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('teth', coins_1.Teth.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('gteth', coins_1.Gteth.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('eth2', coins_1.Eth2.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('teth2', coins_1.Teth2.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('etc', coins_1.Etc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tetc', coins_1.Tetc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('rbtc', coins_1.Rbtc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('trbtc', coins_1.Trbtc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('celo', coins_1.Celo.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tcelo', coins_1.Tcelo.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('avaxc', coins_1.AvaxC.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tavaxc', coins_1.TavaxC.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('xrp', coins_1.Xrp.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('txrp', coins_1.Txrp.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('xlm', coins_1.Xlm.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('txlm', coins_1.Txlm.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('dash', coins_1.Dash.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tdash', coins_1.Tdash.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('zec', coins_1.Zec.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tzec', coins_1.Tzec.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('algo', coins_1.Algo.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('talgo', coins_1.Talgo.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('trx', coins_1.Trx.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('ttrx', coins_1.Ttrx.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('xtz', coins_1.Xtz.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('txtz', coins_1.Txtz.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('hbar', coins_1.Hbar.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('thbar', coins_1.Thbar.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('ofc', coins_1.Ofc.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('susd', coins_1.Susd.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tsusd', coins_1.Tsusd.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('cspr', coins_1.Cspr.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tcspr', coins_1.Tcspr.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('stx', coins_1.Stx.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tstx', coins_1.Tstx.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('sol', coins_1.Sol.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tsol', coins_1.Tsol.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('near', near_1.Near.createInstance);
exports.GlobalCoinFactory.registerCoinConstructor('tnear', tnear_1.TNear.createInstance);
for (const token of [...config_1.tokens.bitcoin.eth.tokens, ...config_1.tokens.testnet.eth.tokens]) {
    const tokenConstructor = coins_1.Erc20Token.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
    exports.GlobalCoinFactory.registerCoinConstructor(token.tokenContractAddress, tokenConstructor);
}
for (const token of [...config_1.tokens.bitcoin.xlm.tokens, ...config_1.tokens.testnet.xlm.tokens]) {
    const tokenConstructor = coins_1.StellarToken.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
}
for (const ofcToken of [...config_1.tokens.bitcoin.ofc.tokens, ...config_1.tokens.testnet.ofc.tokens]) {
    const tokenConstructor = coins_1.OfcToken.createTokenConstructor(ofcToken);
    exports.GlobalCoinFactory.registerCoinConstructor(ofcToken.type, tokenConstructor);
}
for (const token of [...config_1.tokens.bitcoin.celo.tokens, ...config_1.tokens.testnet.celo.tokens]) {
    const tokenConstructor = coins_1.CeloToken.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
    exports.GlobalCoinFactory.registerCoinConstructor(token.tokenContractAddress, tokenConstructor);
}
for (const token of [...config_1.tokens.bitcoin.eos.tokens, ...config_1.tokens.testnet.eos.tokens]) {
    const tokenConstructor = coins_1.EosToken.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
    exports.GlobalCoinFactory.registerCoinConstructor(token.tokenContractAddress, tokenConstructor);
}
for (const token of [...config_1.tokens.bitcoin.algo.tokens, ...config_1.tokens.testnet.algo.tokens]) {
    const tokenConstructor = coins_1.AlgoToken.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
    if (token.alias) {
        exports.GlobalCoinFactory.registerCoinConstructor(token.alias, tokenConstructor);
    }
}
for (const token of [...config_1.tokens.bitcoin.avaxc.tokens, ...config_1.tokens.testnet.avaxc.tokens]) {
    const tokenConstructor = coins_1.AvaxCToken.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
    exports.GlobalCoinFactory.registerCoinConstructor(token.tokenContractAddress, tokenConstructor);
}
for (const token of [...config_1.tokens.bitcoin.fiat.tokens, ...config_1.tokens.testnet.fiat.tokens]) {
    const tokenConstructor = coins_1.FiatToken.createTokenConstructor(token);
    exports.GlobalCoinFactory.registerCoinConstructor(token.type, tokenConstructor);
}
