// bitgo/dist/src/v2/markets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Markets = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const { validateParams } = sdk_core_1.common;
class Markets {
    constructor(bitgo, baseCoin) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
    }
    /**
     * Get the latest price data
     * @param params {}
     * @returns {*} an object containing price and volume data from the
     * current day in a number of currencies
     **/
    async latest(params) {
        return await this.bitgo.get(this.baseCoin.url('/market/latest')).result();
    }
    /**
     * Get yesterday's price data
     * @param params {}
     * @returns {*} an object containing price and volume data from the
     * previous day in a number of currencies
     */
    async yesterday(params) {
        return await this.bitgo.get(this.baseCoin.url('/market/yesterday')).result();
    }
    /**
     * Get price data from up to 90 days prior to today
     * @param params { currencyName: the code for the desired currency, for example USD }
     * @returns {*} an object containing average prices from a number of previous days
     */
    async lastDays(params) {
        validateParams(params, ['currencyName'], []);
        const days = params.days && !isNaN(parseInt(params.days, 10)) ? parseInt(params.days, 10) : 90;
        if (days && days < 0) {
            throw new Error('must use a non-negative number of days');
        }
        return this.bitgo.get(this.baseCoin.url('/market/last/' + days + '/' + params.currencyName)).result();
    }
}
exports.Markets = Markets;
