// bitgo/dist/src/v2/webhooks.js
"use strict";
/**
 * User webhook handler for v2 coins and tokens
 *
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = void 0;
const _ = require("lodash");
const sdk_core_1 = require("@bitgo/sdk-core");
const { validateParams } = sdk_core_1.common;
class Webhooks {
    constructor(bitgo, baseCoin) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
    }
    /**
     * Fetch list of user webhooks
     *
     * @returns {*}
     */
    async list() {
        return await this.bitgo.get(this.baseCoin.url('/webhooks')).result();
    }
    /**
     * Add new user webhook
     *
     * @param params
     * @returns {*}
     */
    async add({ url, type, label, numConfirmations = 0 }) {
        validateParams({ url, type, label, numConfirmations }, ['url', 'type'], ['string', 'numConfirmations']);
        return await this.bitgo.post(this.baseCoin.url('/webhooks')).send({ url, type, label, numConfirmations }).result();
    }
    /**
     * Remove user webhook
     *
     * @param params
     * @returns {*}
     */
    async remove(params) {
        validateParams(params, ['url', 'type'], []);
        return this.bitgo.del(this.baseCoin.url('/webhooks')).send(params).result();
    }
    /**
     * Fetch list of webhook notifications for the user
     *
     * @param params
     * @returns {*}
     */
    async listNotifications(params = {}) {
        const queryProperties = [];
        if (params.prevId) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            queryProperties.push('prevId');
        }
        if (params.limit) {
            if (!_.isNumber(params.limit)) {
                throw new Error('invalid limit argument, expecting number');
            }
            queryProperties.push('limit');
        }
        const query = _.pick(params, queryProperties);
        return this.bitgo.get(this.baseCoin.url('/webhooks/notifications')).query(query).result();
    }
    /**
     * Simulate a user webhook
     *
     * @param params
     * @returns {*}
     */
    async simulate(params) {
        validateParams(params, ['webhookId', 'blockId'], []);
        const webhookId = params.webhookId;
        return this.bitgo
            .post(this.baseCoin.url('/webhooks/' + webhookId + '/simulate'))
            .send(params)
            .result();
    }
}
exports.Webhooks = Webhooks;
