// @bitgo/account-lib/dist/src/coin/eth/transferBuilder/transferBuilderERC721.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC721TransferBuilder = void 0;
var errors_1 = require("../../baseCoin/errors");
var contractCall_1 = require("../contractCall");
var utils_1 = require("../utils");
var baseNFTTransferBuilder_1 = require("./baseNFTTransferBuilder");
var walletUtil_1 = require("../walletUtil");
var utils_2 = require("ethers/lib/utils");
var ERC721TransferBuilder = /** @class */ (function (_super) {
    __extends(ERC721TransferBuilder, _super);
    function ERC721TransferBuilder(serializedData) {
        var _this = _super.call(this, serializedData) || this;
        _this.bytes(0);
        if (serializedData) {
            _this.decodeTransferData(serializedData);
        }
        return _this;
    }
    ERC721TransferBuilder.prototype.tokenContractAddress = function (address) {
        if (utils_1.isValidEthAddress(address)) {
            this._tokenContractAddress = address;
            return this;
        }
        throw new errors_1.InvalidParameterValueError('Invalid address');
    };
    ERC721TransferBuilder.prototype.tokenId = function (token) {
        this._tokenId = token;
        return this;
    };
    ERC721TransferBuilder.prototype.bytes = function (bytesInNumber) {
        this._bytes = utils_2.hexZeroPad(utils_2.hexlify(bytesInNumber), 32);
        return this;
    };
    ERC721TransferBuilder.prototype.signAndBuild = function () {
        if (this.hasMandatoryFields()) {
            var types = ['address', 'address', 'uint256', 'bytes'];
            var values = [this._fromAddress, this._toAddress, this._tokenId, this._bytes];
            var contractCall = new contractCall_1.ContractCall(walletUtil_1.ERC721SafeTransferTypeMethodId, types, values);
            this._data = contractCall.serialize();
            return utils_1.sendMultiSigData(this._tokenContractAddress, // to
            '0', // dummy amount value
            this._data, this._expirationTime, this._sequenceId, this.getSignature());
        }
        throw new errors_1.BuildTransactionError("Missing transfer mandatory fields. \n       Destination (to) address, Source (from) address, sequenceID, the token contract address and tokenID are mandatory");
    };
    ERC721TransferBuilder.prototype.hasMandatoryFields = function () {
        return (this._tokenId !== undefined &&
            this._toAddress !== undefined &&
            this._fromAddress !== undefined &&
            this._tokenContractAddress !== undefined &&
            this._sequenceId !== undefined);
    };
    ERC721TransferBuilder.prototype.decodeTransferData = function (data) {
        var transferData = utils_1.decodeERC721TransferData(data);
        this._toAddress = transferData.to;
        this._fromAddress = transferData.from;
        this._expirationTime = transferData.expireTime;
        this._sequenceId = transferData.sequenceId;
        this._signature = transferData.signature;
        this._tokenContractAddress = transferData.tokenContractAddress;
        this._tokenId = transferData.tokenId;
        if (transferData.data) {
            this._data = transferData.data;
        }
    };
    return ERC721TransferBuilder;
}(baseNFTTransferBuilder_1.BaseNFTTransferBuilder));
exports.ERC721TransferBuilder = ERC721TransferBuilder;
