// @bitgo/account-lib/dist/src/coin/dot/addressInitializationBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddressInitializationBuilder = void 0;
var txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var errors_1 = require("../baseCoin/errors");
var transactionBuilder_1 = require("./transactionBuilder");
var baseCoin_1 = require("../baseCoin");
var iface_1 = require("./iface");
var txnSchema_1 = require("./txnSchema");
var AddressInitializationBuilder = /** @class */ (function (_super) {
    __extends(AddressInitializationBuilder, _super);
    function AddressInitializationBuilder(_coinConfig) {
        var _this = _super.call(this, _coinConfig) || this;
        _this._index = 0;
        return _this;
    }
    /** @inheritDoc */
    AddressInitializationBuilder.prototype.buildTransaction = function () {
        if (this._delegate) {
            return this.buildAddProxyTransaction();
        }
        else {
            return this.buildAnonymousProxyTransaction();
        }
    };
    /**
     * Register a proxy account for the sender that is able to make calls on its behalf.
     *
     * @returns {UnsignedTransaction} an unsigned Dot transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#proxy
     */
    AddressInitializationBuilder.prototype.buildAddProxyTransaction = function () {
        var baseTxInfo = this.createBaseTxInfo();
        return txwrapper_polkadot_1.methods.proxy.addProxy({
            delegate: this._delegate,
            proxyType: this._proxyType,
            delay: this._delay,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    };
    /**
     * Spawn a receive address for the sender
     *
     * @return {UnsignedTransaction} an unsigned Dot transaction
     */
    AddressInitializationBuilder.prototype.buildAnonymousProxyTransaction = function () {
        var baseTxInfo = this.createBaseTxInfo();
        return txwrapper_polkadot_1.methods.proxy.anonymous({
            proxyType: this._proxyType,
            index: this._index,
            delay: parseInt(this._delay, 10),
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    };
    Object.defineProperty(AddressInitializationBuilder.prototype, "transactionType", {
        get: function () {
            return baseCoin_1.TransactionType.AddressInitialization;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * The account to delegate auth to.
     *
     * @param {BaseAddress} owner
     * @returns {AddressInitializationBuilder} This builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-proxies#why-use-a-proxy
     */
    AddressInitializationBuilder.prototype.owner = function (owner) {
        this.validateAddress({ address: owner.address });
        this._delegate = owner.address;
        return this;
    };
    /**
     * Used for disambiguation if multiple calls are made in the same transaction
     * Use 0 as a default
     *
     * @param {number} index
     *
     * @returns {AddressInitializationBuilder} This transfer builder.
     */
    AddressInitializationBuilder.prototype.index = function (index) {
        this.validateValue(new bignumber_js_1.default(index));
        this._index = index;
        return this;
    };
    /**
     * The proxy type to add.
     *
     * @param {proxyType} proxyType
     * @returns {AddressInitializationBuilder} This builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-proxies#proxy-types
     */
    AddressInitializationBuilder.prototype.type = function (proxyType) {
        this._proxyType = proxyType;
        return this;
    };
    /**
     * The number of blocks that an announcement must be in place for.
     * before the corresponding call may be dispatched.
     * If zero, then no announcement is needed.
     * TODO: move to the validity window method once it has been standardized
     *
     * @param {string} delay
     * @returns {AddressInitializationBuilder} This transfer builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-proxies#time-delayed-proxies
     */
    AddressInitializationBuilder.prototype.delay = function (delay) {
        this.validateValue(new bignumber_js_1.default(parseInt(delay, 10)));
        this._delay = delay;
        return this;
    };
    /** @inheritdoc */
    AddressInitializationBuilder.prototype.validateDecodedTransaction = function (decodedTxn) {
        var _a, _b;
        var validationResult;
        if (((_a = decodedTxn.method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.AddProxy) {
            var txMethod = decodedTxn.method.args;
            validationResult = this.validateAddProxyFields(txMethod.delegate, txMethod.proxyType, txMethod.delay);
        }
        else if (((_b = decodedTxn.method) === null || _b === void 0 ? void 0 : _b.name) === iface_1.MethodNames.Anonymous) {
            var txMethod = decodedTxn.method.args;
            validationResult = this.validateAnonymousProxyFields(parseInt(txMethod.index, 10), txMethod.proxyType, txMethod.delay);
        }
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("Transaction validation failed: " + validationResult.error.message);
        }
    };
    /** @inheritdoc */
    AddressInitializationBuilder.prototype.fromImplementation = function (rawTransaction) {
        var _a, _b, _c;
        var tx = _super.prototype.fromImplementation.call(this, rawTransaction);
        if (((_a = this._method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.AddProxy) {
            var txMethod = this._method.args;
            this.owner({ address: txMethod.delegate });
            this.type(txMethod.proxyType);
            this.delay(new bignumber_js_1.default(txMethod.delay).toString());
        }
        else if (((_b = this._method) === null || _b === void 0 ? void 0 : _b.name) === iface_1.MethodNames.Anonymous) {
            var txMethod = this._method.args;
            this.index(new bignumber_js_1.default(txMethod.index).toNumber());
            this.type(txMethod.proxyType);
            this.delay(new bignumber_js_1.default(txMethod.delay).toString());
        }
        else {
            throw new errors_1.InvalidTransactionError("Invalid Transaction Type: " + ((_c = this._method) === null || _c === void 0 ? void 0 : _c.name) + ". Expected " + iface_1.MethodNames.AddProxy + " or " + iface_1.MethodNames.Anonymous);
        }
        return tx;
    };
    /** @inheritdoc */
    AddressInitializationBuilder.prototype.validateTransaction = function (_) {
        _super.prototype.validateTransaction.call(this, _);
        this.validateFields();
    };
    AddressInitializationBuilder.prototype.validateFields = function () {
        var validationResult;
        if (this._delegate) {
            validationResult = this.validateAddProxyFields(this._delegate, this._proxyType, this._delay);
        }
        else {
            validationResult = this.validateAnonymousProxyFields(this._index, this._proxyType, this._delay);
        }
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("AddressInitialization Transaction validation failed: " + validationResult.error.message);
        }
    };
    AddressInitializationBuilder.prototype.validateAddProxyFields = function (delegate, proxyType, delay) {
        return txnSchema_1.AddressInitializationSchema.validate({
            delegate: delegate,
            proxyType: proxyType,
            delay: delay,
        });
    };
    AddressInitializationBuilder.prototype.validateAnonymousProxyFields = function (index, proxyType, delay) {
        return txnSchema_1.AnonymousAddressInitializationSchema.validate({
            proxyType: proxyType,
            index: index,
            delay: delay,
        });
    };
    return AddressInitializationBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.AddressInitializationBuilder = AddressInitializationBuilder;
