// @bitgo/account-lib/dist/src/coin/dot/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionTypes = exports.ProxyType = exports.MethodNames = void 0;
/**
 * Method names for the transaction method. Names change based on the type of transaction e.g 'bond' for the staking transaction
 */
var MethodNames;
(function (MethodNames) {
    MethodNames["AddProxy"] = "addProxy";
    MethodNames["Proxy"] = "proxy";
    MethodNames["Bond"] = "bond";
    MethodNames["TransferKeepAlive"] = "transferKeepAlive";
    MethodNames["Unbond"] = "unbond";
    MethodNames["Anonymous"] = "anonymous";
    MethodNames["Batch"] = "batch";
    MethodNames["Chill"] = "chill";
    MethodNames["WithdrawUnbonded"] = "withdrawUnbonded";
    MethodNames["PayoutStakers"] = "payoutStakers";
})(MethodNames = exports.MethodNames || (exports.MethodNames = {}));
/**
 * The types of proxies that can be setup and used
 * https://wiki.polkadot.network/docs/learn-proxies#proxy-types
 */
var ProxyType;
(function (ProxyType) {
    ProxyType["ANY"] = "Any";
    ProxyType["NON_TRANSFER"] = "NonTransfer";
    ProxyType["STAKING"] = "Staking";
    ProxyType["IDENTTITY_JUDGEMENT"] = "IdentityJudgement";
    ProxyType["CANCEL_PROXY"] = "CancelProxy";
})(ProxyType = exports.ProxyType || (exports.ProxyType = {}));
var TransactionTypes;
(function (TransactionTypes) {
    TransactionTypes["TRANSFER"] = "transfer";
    TransactionTypes["STAKING"] = "staking";
    TransactionTypes["ADDR_INIT"] = "addressInitialization";
    TransactionTypes["UNSTAKING"] = "unstaking";
})(TransactionTypes = exports.TransactionTypes || (exports.TransactionTypes = {}));
