// @bitgo/account-lib/dist/src/coin/dot/keyPair.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
var keyring_1 = require("@polkadot/keyring");
var pair_1 = require("@polkadot/keyring/pair");
var baseCoin_1 = require("../baseCoin");
var utils_1 = __importDefault(require("./utils"));
var crypto_1 = require("../../utils/crypto");
var bs58_1 = __importDefault(require("bs58"));
var TYPE = 'ed25519';
var MAINNET_FORMAT = 0;
var keyring = new keyring_1.Keyring({ type: TYPE });
var KeyPair = /** @class */ (function (_super) {
    __extends(KeyPair, _super);
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    function KeyPair(source) {
        return _super.call(this, source) || this;
    }
    /**
     * Helper function to create the KeyringPair for signing a dot transaction.
     *
     * @returns {KeyringPair} dot KeyringPair
     *
     * @see https://polkadot.js.org/docs/api/start/keyring
     */
    KeyPair.prototype.createPolkadotPair = function () {
        var secretKey = this.keyPair.prv ? new Uint8Array(Buffer.from(this.keyPair.prv, 'hex')) : undefined;
        var publicKey = new Uint8Array(Buffer.from(this.keyPair.pub, 'hex'));
        return pair_1.createPair({ toSS58: keyring.encodeAddress, type: TYPE }, { secretKey: secretKey, publicKey: publicKey });
    };
    /**
     // https://wiki.polkadot.network/docs/learn-accounts#address-format
     * Returns the address in either mainnet polkadot format (starts with 1)
     * or substrate format used for westend (starts with 5)
     */
    KeyPair.prototype.getAddress = function (format) {
        var substrateAddress = this.createPolkadotPair().address;
        // generate polkadot (mainnet) address format
        if (format && format === baseCoin_1.AddressFormat.polkadot) {
            return keyring.encodeAddress(substrateAddress, MAINNET_FORMAT);
        }
        return substrateAddress;
    };
    /** @inheritdoc */
    KeyPair.prototype.getKeys = function () {
        var result = { pub: this.keyPair.pub };
        if (this.keyPair.prv) {
            result.prv = this.keyPair.prv;
        }
        return result;
    };
    /** @inheritdoc */
    KeyPair.prototype.recordKeysFromPrivateKeyInProtocolFormat = function (prv) {
        var decodedSeed = utils_1.default.decodeSeed(prv);
        var bufferFromSeed = Buffer.from(decodedSeed.seed);
        return utils_1.default.keyPairFromSeed(bufferFromSeed).keyPair;
    };
    /** @inheritdoc */
    KeyPair.prototype.recordKeysFromPublicKeyInProtocolFormat = function (pub) {
        var publicKey = keyring.addFromPair({
            // tss common pub is in base58 format and decodes to length of 32
            publicKey: crypto_1.isBase58(pub, 32) ? new Uint8Array(bs58_1.default.decode(pub)) : new Uint8Array(Buffer.from(pub, 'hex')),
            secretKey: new Uint8Array(),
        }).publicKey;
        return { pub: crypto_1.toHex(publicKey) };
    };
    return KeyPair;
}(baseCoin_1.Ed25519KeyPair));
exports.KeyPair = KeyPair;
