// @bitgo/account-lib/dist/src/coin/dot/transferBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
var txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var errors_1 = require("../baseCoin/errors");
var transactionBuilder_1 = require("./transactionBuilder");
var baseCoin_1 = require("../baseCoin");
var iface_1 = require("./iface");
var txnSchema_1 = require("./txnSchema");
var utils_1 = __importDefault(require("./utils"));
var singletonRegistry_1 = require("./singletonRegistry");
var TransferBuilder = /** @class */ (function (_super) {
    __extends(TransferBuilder, _super);
    function TransferBuilder(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    /**
     *
     * Dispatch the given call from an account that the sender is authorised for through add_proxy.
     *
     * @returns {UnsignedTransaction} an unsigned Dot transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#proxy
     */
    TransferBuilder.prototype.buildTransaction = function () {
        var baseTxInfo = this.createBaseTxInfo();
        var transferTx = txwrapper_polkadot_1.methods.balances.transferKeepAlive({
            value: this._amount,
            dest: this._to,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
        if (!this._owner) {
            return transferTx;
        }
        return txwrapper_polkadot_1.methods.proxy.proxy({
            real: this._owner,
            forceProxyType: this._forceProxyType,
            call: transferTx.method,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    };
    Object.defineProperty(TransferBuilder.prototype, "transactionType", {
        get: function () {
            return baseCoin_1.TransactionType.Send;
        },
        enumerable: false,
        configurable: true
    });
    /**
     *
     * The amount for transfer transaction.
     *
     * @param {string} amount
     * @returns {TransferBuilder} This transfer builder.
     *
     * @see https://wiki.polkadot.network/docs/build-protocol-info
     */
    TransferBuilder.prototype.amount = function (amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    };
    /**
     *
     * The destination address for transfer transaction.
     *
     * @param {string} dest
     * @returns {TransferBuilder} This transfer builder.
     *
     * @see https://wiki.polkadot.network/docs/build-protocol-info
     */
    TransferBuilder.prototype.to = function (_a) {
        var address = _a.address;
        this.validateAddress({ address: address });
        this._to = address;
        return this;
    };
    /**
     *
     * The real address of the original tx
     *
     * @param {BaseAddress} real
     * @returns {TransferBuilder} This builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-proxies#why-use-a-proxy
     */
    TransferBuilder.prototype.owner = function (owner) {
        this.validateAddress({ address: owner.address });
        this._owner = owner.address;
        return this;
    };
    /**
     *
     * The proxy type to execute
     *
     * @param {proxyType} forceProxyType
     * @returns {TransferBuilder} This builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-proxies#proxy-types
     */
    TransferBuilder.prototype.forceProxyType = function (forceProxyType) {
        this._forceProxyType = forceProxyType;
        return this;
    };
    /** @inheritdoc */
    TransferBuilder.prototype.validateDecodedTransaction = function (decodedTxn, rawTransaction) {
        var _a, _b;
        if (((_a = decodedTxn.method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.TransferKeepAlive) {
            var txMethod = decodedTxn.method.args;
            var amount = "" + txMethod.value;
            var to = txMethod.dest.id;
            var validationResult = txnSchema_1.TransferTransactionSchema.validate({ amount: amount, to: to });
            if (validationResult.error) {
                throw new errors_1.InvalidTransactionError("Transfer Transaction validation failed: " + validationResult.error.message);
            }
        }
        else if (((_b = decodedTxn.method) === null || _b === void 0 ? void 0 : _b.name) === iface_1.MethodNames.Proxy) {
            var txMethod = decodedTxn.method.args;
            var real = txMethod.real;
            var forceProxyType = txMethod.forceProxyType;
            var decodedCall = utils_1.default.decodeCallMethod(rawTransaction, {
                registry: singletonRegistry_1.SingletonRegistry.getInstance(this._material),
                metadataRpc: this._material.metadata,
            });
            var amount = "" + decodedCall.value;
            var to = decodedCall.dest.id;
            var validationResult = txnSchema_1.ProxyTransactionSchema.validate({ real: real, forceProxyType: forceProxyType, amount: amount, to: to });
            if (validationResult.error) {
                throw new errors_1.InvalidTransactionError("Proxy Transaction validation failed: " + validationResult.error.message);
            }
        }
    };
    /** @inheritdoc */
    TransferBuilder.prototype.fromImplementation = function (rawTransaction) {
        var _a, _b, _c, _d;
        var tx = _super.prototype.fromImplementation.call(this, rawTransaction);
        if (((_a = this._method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.TransferKeepAlive) {
            var txMethod = this._method.args;
            this.amount(txMethod.value);
            this.to({ address: utils_1.default.decodeDotAddress(txMethod.dest.id) });
        }
        else if (((_b = this._method) === null || _b === void 0 ? void 0 : _b.name) === iface_1.MethodNames.Proxy) {
            var txMethod = this._method.args;
            this.owner({ address: utils_1.default.decodeDotAddress(txMethod.real) });
            this.forceProxyType(txMethod.forceProxyType);
            var decodedCall = utils_1.default.decodeCallMethod(rawTransaction, {
                registry: singletonRegistry_1.SingletonRegistry.getInstance(this._material),
                metadataRpc: this._material.metadata,
            });
            if (!decodedCall.value || !decodedCall.dest) {
                throw new errors_1.InvalidTransactionError("Invalid Proxy Transaction Method: " + ((_c = this._method) === null || _c === void 0 ? void 0 : _c.name) + ". Expected transferKeepAlive");
            }
            this.amount("" + decodedCall.value);
            this.to({ address: utils_1.default.decodeDotAddress(decodedCall.dest.id) });
        }
        else {
            throw new errors_1.InvalidTransactionError("Invalid Transaction Type: " + ((_d = this._method) === null || _d === void 0 ? void 0 : _d.name) + ". Expected a transferKeepAlive or a proxy transferKeepAlive transaction");
        }
        return tx;
    };
    /** @inheritdoc */
    TransferBuilder.prototype.validateTransaction = function (_) {
        _super.prototype.validateTransaction.call(this, _);
        this.validateFields(this._to, this._amount, this._owner, this._forceProxyType);
    };
    TransferBuilder.prototype.validateFields = function (to, amount, real, forceProxyType) {
        var validationResult = forceProxyType
            ? txnSchema_1.ProxyTransactionSchema.validate({ to: to, amount: amount, real: real, forceProxyType: forceProxyType })
            : txnSchema_1.TransferTransactionSchema.validate({ amount: amount, to: to });
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("Proxy/TransferKeepAlive Transaction validation failed: " + validationResult.error.message);
        }
    };
    return TransferBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.TransferBuilder = TransferBuilder;
