// @bitgo/account-lib/dist/src/coin/dot/txnSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaimTransactionSchema = exports.WithdrawUnstakedTransactionSchema = exports.UnstakeTransactionSchema = exports.ProxyTransactionSchema = exports.BatchTransactionSchema = exports.AnonymousAddressInitializationSchema = exports.AddressInitializationSchema = exports.StakeTransactionSchema = exports.TransferTransactionSchema = exports.SignedTransactionSchema = exports.SigningPayloadTransactionSchema = exports.BaseTransactionSchema = void 0;
var joi_1 = __importDefault(require("joi"));
var utils_1 = __importDefault(require("./utils"));
var addressSchema = joi_1.default.string().custom(function (addr) { return utils_1.default.isValidAddress(addr); });
var proxyTypes = [
    'Any',
    'NonTransfer',
    'Governance',
    'Staking',
    'UnusedSudoBalances',
    'IdentityJudgement',
    'CancelProxy',
];
exports.BaseTransactionSchema = joi_1.default.object({
    sender: addressSchema.required(),
    blockNumber: joi_1.default.number().required(),
    blockHash: joi_1.default.string().required(),
    genesisHash: joi_1.default.string().required(),
    specVersion: joi_1.default.number().required(),
    specName: joi_1.default.string().valid('kusama', 'polkadot', 'westend', 'statemint', 'statemine').required(),
    transactionVersion: joi_1.default.number().required(),
    chainName: joi_1.default.string().required(),
    eraPeriod: joi_1.default.number().required(),
    nonce: joi_1.default.number().required(),
    tip: joi_1.default.number().optional(),
});
exports.SigningPayloadTransactionSchema = joi_1.default.object({
    eraPeriod: joi_1.default.number().optional(),
    blockHash: joi_1.default.string().required(),
    nonce: joi_1.default.number().required(),
    tip: joi_1.default.number().optional(),
});
exports.SignedTransactionSchema = joi_1.default.object({
    sender: addressSchema.required(),
    nonce: joi_1.default.number().required(),
    eraPeriod: joi_1.default.number().optional(),
    tip: joi_1.default.number().optional(),
});
exports.TransferTransactionSchema = joi_1.default.object({
    amount: joi_1.default.string().required(),
    to: addressSchema.required(),
});
exports.StakeTransactionSchema = joi_1.default.object({
    value: joi_1.default.string().required(),
    controller: addressSchema.required(),
    payee: [
        joi_1.default.string(),
        joi_1.default.object({
            account: joi_1.default.string().optional(),
            controller: joi_1.default.equal(null).optional(),
            staked: joi_1.default.equal(null).optional(),
            stash: joi_1.default.equal(null).optional(),
        }),
        joi_1.default.object({
            Account: joi_1.default.string().required(),
        }),
    ],
});
exports.AddressInitializationSchema = joi_1.default.object({
    proxyType: (_a = joi_1.default
        .string())
        .valid.apply(_a, proxyTypes).required(),
    delegate: addressSchema.required(),
    delay: joi_1.default.string().required(),
});
exports.AnonymousAddressInitializationSchema = joi_1.default.object({
    proxyType: (_b = joi_1.default
        .string())
        .valid.apply(_b, proxyTypes).required(),
    index: joi_1.default.number().required(),
    delay: joi_1.default.number().required(),
});
exports.BatchTransactionSchema = joi_1.default.object({
    calls: joi_1.default
        .alternatives()
        .try(joi_1.default.array().items(joi_1.default.string()), joi_1.default.array().items(joi_1.default.object({ callIndex: joi_1.default.string(), args: joi_1.default.object() }))),
});
exports.ProxyTransactionSchema = joi_1.default.object({
    real: addressSchema.required(),
    forceProxyType: (_c = joi_1.default
        .string())
        .valid.apply(_c, proxyTypes).required(),
    amount: joi_1.default.string().required(),
    to: addressSchema.required(),
});
exports.UnstakeTransactionSchema = joi_1.default.object({
    value: joi_1.default.string().required(),
});
exports.WithdrawUnstakedTransactionSchema = joi_1.default.object({
    value: joi_1.default.number().required(),
});
exports.ClaimTransactionSchema = joi_1.default.object({
    claimEra: joi_1.default.string().required(),
    validatorStash: addressSchema.required(),
});
