// @bitgo/account-lib/dist/src/coin/dot/unstakeBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnstakeBuilder = void 0;
var txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var baseCoin_1 = require("../baseCoin");
var errors_1 = require("../baseCoin/errors");
var iface_1 = require("./iface");
var transactionBuilder_1 = require("./transactionBuilder");
var txnSchema_1 = require("./txnSchema");
var UnstakeBuilder = /** @class */ (function (_super) {
    __extends(UnstakeBuilder, _super);
    function UnstakeBuilder(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    /**
     *
     * Take the origin account as a stash and lock up value of its balance.
     * Controller will be the account that controls it.
     *
     * @returns {UnsignedTransaction} an unsigned Dot transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#staking
     */
    UnstakeBuilder.prototype.buildTransaction = function () {
        var baseTxInfo = this.createBaseTxInfo();
        return txwrapper_polkadot_1.methods.staking.unbond({
            value: this._amount,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    };
    Object.defineProperty(UnstakeBuilder.prototype, "transactionType", {
        get: function () {
            return baseCoin_1.TransactionType.StakingUnlock;
        },
        enumerable: false,
        configurable: true
    });
    /**
     *
     * The amount to unstake.
     *
     * @param {string} amount
     * @returns {StakeBuilder} This stake builder.
     *
     * @see https://wiki.polkadot.network/docs/learn-nominator#required-minimum-stake
     */
    UnstakeBuilder.prototype.amount = function (amount) {
        this.validateValue(new bignumber_js_1.default(amount));
        this._amount = amount;
        return this;
    };
    /** @inheritdoc */
    UnstakeBuilder.prototype.validateDecodedTransaction = function (decodedTxn) {
        var _a;
        if (((_a = decodedTxn.method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.Unbond) {
            var txMethod = decodedTxn.method.args;
            var value = txMethod.value;
            var validationResult = txnSchema_1.UnstakeTransactionSchema.validate({ value: value });
            if (validationResult.error) {
                throw new errors_1.InvalidTransactionError("Unstake Transaction validation failed: " + validationResult.error.message);
            }
        }
    };
    /** @inheritdoc */
    UnstakeBuilder.prototype.fromImplementation = function (rawTransaction) {
        var _a, _b;
        var tx = _super.prototype.fromImplementation.call(this, rawTransaction);
        if (((_a = this._method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.Unbond) {
            var txMethod = this._method.args;
            this.amount(txMethod.value);
        }
        else {
            throw new errors_1.InvalidTransactionError("Invalid Transaction Type: " + ((_b = this._method) === null || _b === void 0 ? void 0 : _b.name) + ". Expected unbond");
        }
        return tx;
    };
    /** @inheritdoc */
    UnstakeBuilder.prototype.validateTransaction = function (_) {
        _super.prototype.validateTransaction.call(this, _);
        this.validateFields(this._amount);
    };
    UnstakeBuilder.prototype.validateFields = function (value) {
        var validationResult = txnSchema_1.UnstakeTransactionSchema.validate({
            value: value,
        });
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("Unstake Builder Transaction validation failed: " + validationResult.error.message);
        }
    };
    return UnstakeBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.UnstakeBuilder = UnstakeBuilder;
