// @bitgo/account-lib/dist/src/coin/dot/withdrawUnstakedBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawUnstakedBuilder = void 0;
var txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var baseCoin_1 = require("../baseCoin");
var errors_1 = require("../baseCoin/errors");
var iface_1 = require("./iface");
var transactionBuilder_1 = require("./transactionBuilder");
var txnSchema_1 = require("./txnSchema");
var WithdrawUnstakedBuilder = /** @class */ (function (_super) {
    __extends(WithdrawUnstakedBuilder, _super);
    function WithdrawUnstakedBuilder(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    /**
     *
     * Frees up any unlocked chunks.
     * The balance can now be used by the stash account to do whatever it wants.
     *
     * @returns {UnsignedTransaction} an unsigned Dot transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#withdrawunbondednum_slashing_spans-u32
     */
    WithdrawUnstakedBuilder.prototype.buildTransaction = function () {
        var baseTxInfo = this.createBaseTxInfo();
        // TODO: The method changes the args type from number to string, verify it doesn't break anything
        return txwrapper_polkadot_1.methods.staking.withdrawUnbonded({
            numSlashingSpans: this._numSlashingSpans,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    };
    Object.defineProperty(WithdrawUnstakedBuilder.prototype, "transactionType", {
        get: function () {
            return baseCoin_1.TransactionType.StakingWithdraw;
        },
        enumerable: false,
        configurable: true
    });
    /**
     *
     * The number of slashing spans.
     *
     * @param {number} slashingSpans
     * @returns {WithdrawUnstakedBuilder} This withdrawUnstaked builder.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#withdrawunbondednum_slashing_spans-u32
     */
    WithdrawUnstakedBuilder.prototype.slashingSpans = function (slashingSpans) {
        this.validateValue(new bignumber_js_1.default(slashingSpans));
        this._numSlashingSpans = slashingSpans;
        return this;
    };
    /** @inheritdoc */
    WithdrawUnstakedBuilder.prototype.validateDecodedTransaction = function (decodedTxn) {
        var _a;
        if (((_a = decodedTxn.method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.WithdrawUnbonded) {
            var txMethod = decodedTxn.method.args;
            var value = txMethod.numSlashingSpans;
            var validationResult = txnSchema_1.WithdrawUnstakedTransactionSchema.validate({ value: value });
            if (validationResult.error) {
                throw new errors_1.InvalidTransactionError("WithdrawUnstaked Transaction validation failed: " + validationResult.error.message);
            }
        }
    };
    /** @inheritdoc */
    WithdrawUnstakedBuilder.prototype.fromImplementation = function (rawTransaction) {
        var _a, _b;
        var tx = _super.prototype.fromImplementation.call(this, rawTransaction);
        if (((_a = this._method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.WithdrawUnbonded) {
            var txMethod = this._method.args;
            this.slashingSpans(txMethod.numSlashingSpans);
        }
        else {
            throw new errors_1.InvalidTransactionError("Invalid Transaction Type: " + ((_b = this._method) === null || _b === void 0 ? void 0 : _b.name) + ". Expected withdrawUnbonded");
        }
        return tx;
    };
    /** @inheritdoc */
    WithdrawUnstakedBuilder.prototype.validateTransaction = function (_) {
        _super.prototype.validateTransaction.call(this, _);
        this.validateFields(this._numSlashingSpans);
    };
    WithdrawUnstakedBuilder.prototype.validateFields = function (value) {
        var validationResult = txnSchema_1.WithdrawUnstakedTransactionSchema.validate({
            value: value,
        });
        if (validationResult.error) {
            throw new errors_1.InvalidTransactionError("WithdrawUnstaked Builder Transaction validation failed: " + validationResult.error.message);
        }
    };
    return WithdrawUnstakedBuilder;
}(transactionBuilder_1.TransactionBuilder));
exports.WithdrawUnstakedBuilder = WithdrawUnstakedBuilder;
