// @bitgo/account-lib/dist/src/coin/hbar/keyPair.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
var sdk_1 = require("@hashgraph/sdk");
var ed25519KeyPair_1 = require("../baseCoin/ed25519KeyPair");
var errors_1 = require("../baseCoin/errors");
var utils_1 = require("./utils");
var PUBLIC_KEY_PREFIX = '302a300506032b6570032100';
var PRIVATE_KEY_PREFIX = '302e020100300506032b657004220420';
var KeyPair = /** @class */ (function (_super) {
    __extends(KeyPair, _super);
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    function KeyPair(source) {
        return _super.call(this, source) || this;
    }
    /**
     * Hedera default keys format is a pair of Uint8Array keys
     *
     * @param {boolean} raw defines if the key is returned in raw or protocol default format
     * @returns { DefaultKeys } The keys in the defined format
     */
    KeyPair.prototype.getKeys = function (raw) {
        if (raw === void 0) { raw = false; }
        var pub = sdk_1.PublicKey.fromString(this.keyPair.pub).toString();
        var result = {
            pub: raw ? pub.slice(PUBLIC_KEY_PREFIX.length) : pub,
        };
        if (this.keyPair.prv) {
            var prv = sdk_1.PrivateKey.fromString(this.keyPair.prv).toString();
            result.prv = raw ? prv.slice(PRIVATE_KEY_PREFIX.length) : prv;
        }
        return result;
    };
    /** @inheritdoc */
    KeyPair.prototype.getAddress = function (format) {
        throw new errors_1.NotSupported('Address derivation is not supported in Hedera');
    };
    /** @inheritdoc */
    KeyPair.prototype.recordKeysFromPublicKeyInProtocolFormat = function (pub) {
        try {
            var hederaPub = sdk_1.PublicKey.fromString(pub.toLowerCase()).toString();
            var ed25519Pub = utils_1.removePrefix(PUBLIC_KEY_PREFIX, hederaPub);
            return { pub: ed25519Pub };
        }
        catch (e) {
            throw new errors_1.InvalidKey('Invalid public key: ' + pub);
        }
    };
    /** @inheritdoc */
    KeyPair.prototype.recordKeysFromPrivateKeyInProtocolFormat = function (prv) {
        if (!/^([a-f0-9]{2})+$/i.test(prv)) {
            throw new errors_1.InvalidKey('Invalid private key length. Must be a hex and multiple of 2');
        }
        try {
            var hederaPrv = sdk_1.PrivateKey.fromString(prv);
            var ed25519Prv = utils_1.removePrefix(PRIVATE_KEY_PREFIX, hederaPrv.toString());
            var ed25519Pub = utils_1.removePrefix(PUBLIC_KEY_PREFIX, hederaPrv.publicKey.toString());
            return {
                prv: ed25519Prv,
                pub: ed25519Pub,
            };
        }
        catch (e) {
            throw new errors_1.InvalidKey('Invalid private key');
        }
    };
    return KeyPair;
}(ed25519KeyPair_1.Ed25519KeyPair));
exports.KeyPair = KeyPair;
