// @bitgo/account-lib/dist/src/coin/hbar/transactionBuilderFactory.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
var errors_1 = require("../baseCoin/errors");
var baseCoin_1 = require("../baseCoin");
var walletInitializationBuilder_1 = require("./walletInitializationBuilder");
var transferBuilder_1 = require("./transferBuilder");
var transaction_1 = require("./transaction");
var utils_1 = require("./utils");
var TransactionBuilderFactory = /** @class */ (function (_super) {
    __extends(TransactionBuilderFactory, _super);
    function TransactionBuilderFactory(_coinConfig) {
        return _super.call(this, _coinConfig) || this;
    }
    /** @inheritdoc */
    TransactionBuilderFactory.prototype.getWalletInitializationBuilder = function (tx) {
        return this.initializeBuilder(tx, new walletInitializationBuilder_1.WalletInitializationBuilder(this._coinConfig));
    };
    /** @inheritDoc */
    TransactionBuilderFactory.prototype.getTransferBuilder = function (tx) {
        return this.initializeBuilder(tx, new transferBuilder_1.TransferBuilder(this._coinConfig));
    };
    /** @inheritDoc */
    TransactionBuilderFactory.prototype.from = function (raw) {
        this.validateRawTransaction(raw);
        var tx = this.parseTransaction(raw);
        switch (tx.txBody.data) {
            case 'cryptoTransfer':
                return this.getTransferBuilder(tx);
            case 'cryptoCreateAccount':
                return this.getWalletInitializationBuilder(tx);
            default:
                throw new errors_1.InvalidTransactionError('Invalid transaction ' + tx.txBody.data);
        }
    };
    /**
     * Initialize the builder with the given transaction
     *
     * @param {Transaction | undefined} tx - the transaction used to initialize the builder
     * @param {TransactionBuilder} builder - the builder to be initialized
     * @returns {TransactionBuilder} the builder initialized
     */
    TransactionBuilderFactory.prototype.initializeBuilder = function (tx, builder) {
        if (tx) {
            builder.initBuilder(tx);
        }
        return builder;
    };
    /**
     * Returns a transaction instance from the encoded value
     *
     * @param {Uint8Array | string} rawTransaction - encoded transaction
     * @returns {Transaction} the parsed transaction instance
     */
    TransactionBuilderFactory.prototype.parseTransaction = function (rawTransaction) {
        var tx = new transaction_1.Transaction(this._coinConfig);
        var buffer;
        if (typeof rawTransaction === 'string') {
            buffer = utils_1.toUint8Array(rawTransaction);
        }
        else {
            buffer = rawTransaction;
        }
        tx.bodyBytes(buffer);
        return tx;
    };
    /**
     * Check the raw transaction has a valid format in the blockchain context, throw otherwise.
     *
     * @param {any} rawTransaction - Transaction in any format
     */
    TransactionBuilderFactory.prototype.validateRawTransaction = function (rawTransaction) {
        if (!utils_1.isValidRawTransactionFormat(rawTransaction)) {
            throw new errors_1.ParseTransactionError('Invalid raw transaction');
        }
    };
    return TransactionBuilderFactory;
}(baseCoin_1.BaseTransactionBuilderFactory));
exports.TransactionBuilderFactory = TransactionBuilderFactory;
