// @substrate/txwrapper-core/lib/test-helpers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolkadotSS58Format = exports.TEST_METHOD_ARGS = exports.KUSAMA_9160_TEST_OPTIONS = exports.POLKADOT_9122_TEST_OPTIONS_CALLS_ONLY = exports.POLKADOT_9122_TEST_OPTIONS = exports.POLKADOT_29_TEST_OPTIONS = exports.POLKADOT_25_TEST_OPTIONS = exports.TEST_BASE_TX_INFO_V29 = exports.TEST_BASE_TX_INFO = exports.metadataRpc = void 0;
const getRegistryKusama_1 = require("./getRegistryKusama");
const getRegistryPolkadot_1 = require("./getRegistryPolkadot");
const kusamaV9160MetadataHex_1 = require("./metadata/kusamaV9160MetadataHex");
const polkadotV9122MetadataHex_1 = require("./metadata/polkadotV9122MetadataHex");
const staticV3_1_1_1 = require("./staticV3-1-1");
Object.defineProperty(exports, "metadataRpc", { enumerable: true, get: function () { return staticV3_1_1_1.metadataRpc; } });
const staticV4_3_1_1 = require("./staticV4-3-1");
/**
 * Base tx information common to all tested transactions
 */
exports.TEST_BASE_TX_INFO = {
    address: '15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5',
    blockHash: '0x1fc7493f3c1e9ac758a183839906475f8363aafb1b1d3e910fe16fab4ae1b582',
    blockNumber: 4302222,
    eraPeriod: 2400,
    genesisHash: '0xe3777fa922cafbff200cadeaea1a76bd7898ad5b89f7848999058b50e715f636',
    metadataRpc: staticV3_1_1_1.metadataRpc,
    nonce: 2,
    specVersion: 1019,
    tip: 0,
    transactionVersion: 6,
};
/**
 * Base tx information common to all tested transactions using runtime v29
 */
exports.TEST_BASE_TX_INFO_V29 = {
    address: '15oF4uVJwmo4TdGW7VfQxNLavjCXviqxT9S1MgbjMNHr6Sp5',
    blockHash: '0x1fc7493f3c1e9ac758a183839906475f8363aafb1b1d3e910fe16fab4ae1b582',
    blockNumber: 4302222,
    eraPeriod: 2400,
    genesisHash: '0xe3777fa922cafbff200cadeaea1a76bd7898ad5b89f7848999058b50e715f636',
    metadataRpc: staticV4_3_1_1.metadataRpcV29,
    nonce: 2,
    specVersion: 1019,
    tip: 0,
    transactionVersion: 6,
};
/**
 * Use this registry in tests.
 */
exports.POLKADOT_25_TEST_OPTIONS = {
    metadataRpc: staticV3_1_1_1.metadataRpc,
    registry: (0, getRegistryPolkadot_1.getRegistryPolkadot)(25, staticV3_1_1_1.metadataRpc),
};
/**
 * Test options for runtime v29
 */
exports.POLKADOT_29_TEST_OPTIONS = {
    metadataRpc: staticV4_3_1_1.metadataRpcV29,
    registry: (0, getRegistryPolkadot_1.getRegistryPolkadot)(29, staticV4_3_1_1.metadataRpcV29),
};
/**
 * Test options for runtime v9122
 */
exports.POLKADOT_9122_TEST_OPTIONS = {
    metadataRpc: polkadotV9122MetadataHex_1.polkadotV9122MetadataHex,
    registry: (0, getRegistryPolkadot_1.getRegistryPolkadot)(9122, polkadotV9122MetadataHex_1.polkadotV9122MetadataHex),
};
/**
 * Test options for runtime v9122 with calls only metadata
 */
exports.POLKADOT_9122_TEST_OPTIONS_CALLS_ONLY = {
    metadataRpc: polkadotV9122MetadataHex_1.polkadotV9122MetadataHex,
    registry: (0, getRegistryPolkadot_1.getRegistryPolkadot)(9122, polkadotV9122MetadataHex_1.polkadotV9122MetadataHex),
    asCallsOnlyArg: true,
};
/**
 * Test options for runtime v9160 Kusama
 */
exports.KUSAMA_9160_TEST_OPTIONS = {
    metadataRpc: kusamaV9160MetadataHex_1.kusamaV9160MetadataHex,
    registry: (0, getRegistryKusama_1.getRegistryKusama)(9160, kusamaV9160MetadataHex_1.kusamaV9160MetadataHex),
};
/**
 * Dummy arguments for all methods we're testing.
 */
exports.TEST_METHOD_ARGS = {
    balances: {
        transfer: {
            dest: '14PiLtTQ3EBwsn6KimaBLH7qsZrryxFZtqFwAognHVZhFxU1',
            value: 12,
        },
    },
};
/**
 * Prefix for ss58-encoded addresses on Polkadot, Kusama, and Westend. Note:
 * 42, the Westend prefix, is also the default for Substrate-based chains.
 */
var PolkadotSS58Format;
(function (PolkadotSS58Format) {
    PolkadotSS58Format[PolkadotSS58Format["polkadot"] = 0] = "polkadot";
    PolkadotSS58Format[PolkadotSS58Format["kusama"] = 2] = "kusama";
    PolkadotSS58Format[PolkadotSS58Format["westend"] = 42] = "westend";
    PolkadotSS58Format[PolkadotSS58Format["substrate"] = 42] = "substrate";
})(PolkadotSS58Format = exports.PolkadotSS58Format || (exports.PolkadotSS58Format = {}));
