// bitgo/dist/src/v2/internal/blsUtils.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlsUtils = void 0;
const crypto_1 = require("crypto");
const openpgp_1 = require("openpgp");
const account_lib_1 = require("@bitgo/account-lib");
const mpcUtils_1 = require("./mpcUtils");
/**
 * Utility functions for BLS-DKG work flows.
 */
class BlsUtils extends mpcUtils_1.MpcUtils {
    constructor(bitgo, baseCoin) {
        super(bitgo, baseCoin);
    }
    /**
     * Creates a Keychain containing the User's BLS-DKG signing materials.
     *
     * @param userGpgKey - ephemeral GPG key to encrypt / decrypt sensitve data exchanged between user and server
     * @param userKeyShare - user's BLS-DKG key share
     * @param backupKeyShare - backup's BLS-DKG key share
     * @param bitgoKeychain - previously created BitGo keychain; must be compatible with user and backup key shares
     * @param passphrase - wallet passphrase used to encrypt user's signing materials
     */
    async createUserKeychain(userGpgKey, userKeyShare, backupKeyShare, bitgoKeychain, passphrase, originalPasscodeEncryptionCode) {
        const bitgoKeyShares = bitgoKeychain.keyShares;
        if (!bitgoKeyShares) {
            throw new Error('Missing BitGo key shares');
        }
        const bitGoToUserShare = bitgoKeyShares.find((keyShare) => keyShare.from === 'bitgo' && keyShare.to === 'user');
        if (!bitGoToUserShare) {
            throw new Error('Missing BitGo to User key share');
        }
        if (!userKeyShare.secretShares || !userKeyShare.pub) {
            throw new Error('Invalid user key shares');
        }
        if (!backupKeyShare.secretShares || !backupKeyShare.pub) {
            throw new Error('Invalid backup key shares');
        }
        const bitGoToUserPrivateShare = await this.decryptPrivateShare(bitGoToUserShare.privateShare, userGpgKey);
        const userPrivateKey = account_lib_1.BaseCoin.BlsKeyPair.aggregatePrvkeys([
            userKeyShare.secretShares[0],
            backupKeyShare.secretShares[0],
            bitGoToUserPrivateShare,
        ]);
        const commonPub = account_lib_1.BaseCoin.BlsKeyPair.aggregatePubkeys([
            userKeyShare.pub,
            backupKeyShare.pub,
            bitGoToUserShare.publicShare,
        ]);
        if (commonPub !== bitgoKeychain.commonPub) {
            throw new Error('Failed to create user keychain - commonPubs do not match.');
        }
        const userKeychainParams = {
            source: 'user',
            type: 'blsdkg',
            commonPub: commonPub,
            encryptedPrv: this.bitgo.encrypt({ input: userPrivateKey, password: passphrase }),
            originalPasscodeEncryptionCode: originalPasscodeEncryptionCode,
        };
        return await this.baseCoin.keychains().add(userKeychainParams);
    }
    /**
     * Creates a Keychain containing the Backup party's BLS-DKG signing materials.
     *
     * @param userGpgKey - ephemeral GPG key to encrypt / decrypt sensitve data exchanged between user and server
     * @param userKeyShare - User's BLS-DKG Keyshare
     * @param backupKeyShare - Backup's BLS-DKG Keyshare
     * @param bitgoKeychain - previously created BitGo keychain; must be compatible with user and backup key shares
     * @param passphrase - wallet passphrase used to encrypt user's signing materials
     */
    async createBackupKeychain(userGpgKey, userKeyShare, backupKeyShare, bitgoKeychain, passphrase) {
        const bitgoKeyShares = bitgoKeychain.keyShares;
        if (!bitgoKeyShares) {
            throw new Error('Invalid bitgo keyshares');
        }
        const bitGoToBackupShare = bitgoKeyShares.find((keyShare) => keyShare.from === 'bitgo' && keyShare.to === 'backup');
        if (!bitGoToBackupShare) {
            throw new Error('Missing BitGo to User key share');
        }
        if (!userKeyShare.secretShares || !userKeyShare.pub) {
            throw new Error('Invalid user key shares');
        }
        if (!backupKeyShare.secretShares || !backupKeyShare.pub) {
            throw new Error('Invalid backup key shares');
        }
        const bitGoToBackupPrivateShare = await this.decryptPrivateShare(bitGoToBackupShare.privateShare, userGpgKey);
        const backupPrivateKey = account_lib_1.BaseCoin.BlsKeyPair.aggregatePrvkeys([
            userKeyShare.secretShares[1],
            backupKeyShare.secretShares[1],
            bitGoToBackupPrivateShare,
        ]);
        const commonPub = account_lib_1.BaseCoin.BlsKeyPair.aggregatePubkeys([
            userKeyShare.pub,
            backupKeyShare.pub,
            bitGoToBackupShare.publicShare,
        ]);
        if (commonPub !== bitgoKeychain.commonPub) {
            throw new Error('Failed to create backup keychain - commonPubs do not match.');
        }
        return await this.baseCoin.keychains().createBackup({
            source: 'backup',
            type: 'blsdkg',
            commonPub: commonPub,
            prv: backupPrivateKey,
            encryptedPrv: this.bitgo.encrypt({ input: backupPrivateKey, password: passphrase }),
        });
    }
    /**
     * Creates a Keychain containing BitGo's BLS-DKG signing materials.
     *
     * @param userGpgKey - ephemeral GPG key to encrypt / decrypt sensitve data exchanged between user and server
     * @param userKeyShare - user's BLS-DKG key share
     * @param backupKeyShare - backup's BLS-DKG key share
     */
    async createBitgoKeychain(userGpgKey, userKeyShare, backupKeyShare, enterprise) {
        if (!userKeyShare.secretShares || !userKeyShare.pub) {
            throw new Error('Invalid user key shares');
        }
        if (!backupKeyShare.secretShares || !backupKeyShare.pub) {
            throw new Error('Invalid backup key shares');
        }
        const userToBitgoKeyShare = {
            publicShare: userKeyShare.pub,
            privateShare: userKeyShare.secretShares[2],
        };
        const backupToBitgoKeyShare = {
            publicShare: backupKeyShare.pub,
            privateShare: backupKeyShare.secretShares[2],
        };
        return await this.createBitgoKeychainInWP(userGpgKey, userToBitgoKeyShare, backupToBitgoKeyShare, 'blsdkg', enterprise);
    }
    /**
     * Creates User, Backup, and BitGo BLS-DKG Keychains.
     *
     * @param params.passphrase - passphrase used to encrypt signing materials created for User and Backup
     */
    async createKeychains(params) {
        const userKeyShare = this.baseCoin.generateKeyPair();
        const backupKeyShare = this.baseCoin.generateKeyPair();
        const randomHexString = crypto_1.randomBytes(12).toString('hex');
        const userGpgKey = await openpgp_1.generateKey({
            userIDs: [
                {
                    name: randomHexString,
                    email: `${randomHexString}@${randomHexString}.com`,
                },
            ],
        });
        const bitgoKeychain = await this.createBitgoKeychain(userGpgKey, userKeyShare, backupKeyShare, params.enterprise);
        const userKeychainPromise = this.createUserKeychain(userGpgKey, userKeyShare, backupKeyShare, bitgoKeychain, params.passphrase, params.originalPasscodeEncryptionCode);
        const backupKeychainPromise = this.createBackupKeychain(userGpgKey, userKeyShare, backupKeyShare, bitgoKeychain, params.passphrase);
        const [userKeychain, backupKeychain] = await Promise.all([userKeychainPromise, backupKeychainPromise]);
        // create wallet
        const keychains = {
            userKeychain,
            backupKeychain,
            bitgoKeychain,
        };
        return keychains;
    }
}
exports.BlsUtils = BlsUtils;
