// bitgo/dist/src/v2/internal/mpcUtils.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MpcUtils = void 0;
const openpgp_1 = require("openpgp");
const opengpgUtils_1 = require("./opengpgUtils");
class MpcUtils {
    constructor(bitgo, baseCoin) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
    }
    async decryptPrivateShare(privateShare, userGpgKey) {
        const privateShareMessage = await openpgp_1.readMessage({
            armoredMessage: privateShare,
        });
        const userGpgPrivateKey = await openpgp_1.readPrivateKey({ armoredKey: userGpgKey.privateKey });
        const decryptedPrivateShare = (await openpgp_1.decrypt({
            message: privateShareMessage,
            decryptionKeys: [userGpgPrivateKey],
            format: 'utf8',
        })).data;
        return decryptedPrivateShare;
    }
    async createBitgoKeychainInWP(userGpgKey, userKeyShare, backupKeyShare, keyType, enterprise) {
        const bitgoKey = await opengpgUtils_1.getBitgoGpgPubKey(this.bitgo);
        const encUserToBitGoMessage = await opengpgUtils_1.encryptText(userKeyShare.privateShare, bitgoKey);
        const encBackupToBitGoMessage = await opengpgUtils_1.encryptText(backupKeyShare.privateShare, bitgoKey);
        const createBitGoMPCParams = {
            type: keyType,
            source: 'bitgo',
            keyShares: [
                {
                    from: 'user',
                    to: 'bitgo',
                    publicShare: userKeyShare.publicShare,
                    privateShare: encUserToBitGoMessage,
                },
                {
                    from: 'backup',
                    to: 'bitgo',
                    publicShare: backupKeyShare.publicShare,
                    privateShare: encBackupToBitGoMessage,
                },
            ],
            userGPGPublicKey: userGpgKey.publicKey,
            backupGPGPublicKey: userGpgKey.publicKey,
            enterprise: enterprise,
        };
        return await this.baseCoin.keychains().add(createBitGoMPCParams);
    }
}
exports.MpcUtils = MpcUtils;
