// bitgo/dist/src/v2/internal/opengpgUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encryptText = exports.getBitgoGpgPubKey = void 0;
const openpgp_1 = require("openpgp");
/**
 * Fetches BitGo's pubic gpg key used in MPC flows
 * @param {BitGo} bitgo BitGo object
 * @return {Key} public gpg key
 */
async function getBitgoGpgPubKey(bitgo) {
    const constants = await bitgo.fetchConstants();
    if (!constants.mpc || !constants.mpc.bitgoPublicKey) {
        throw new Error('Unable to create MPC keys - bitgoPublicKey is missing from constants');
    }
    const bitgoPublicKeyStr = constants.mpc.bitgoPublicKey;
    return await openpgp_1.readKey({ armoredKey: bitgoPublicKeyStr });
}
exports.getBitgoGpgPubKey = getBitgoGpgPubKey;
/**
 * Encrypts string using gpg key
 * @param text string to encrypt
 * @param key encryption key
 * @return {string} encrypted string
 */
async function encryptText(text, key) {
    const messageToEncrypt = await openpgp_1.createMessage({
        text,
    });
    return await openpgp_1.encrypt({
        message: messageToEncrypt,
        encryptionKeys: [key],
        format: 'armored',
        config: {
            rejectCurves: new Set(),
            showVersion: false,
            showComment: false,
        },
    });
}
exports.encryptText = encryptText;
