// bitgo/dist/src/v2/trading/affirmation.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Affirmation = exports.AffirmationStatus = void 0;
var AffirmationStatus;
(function (AffirmationStatus) {
    AffirmationStatus["PENDING"] = "pending";
    AffirmationStatus["OVERDUE"] = "overdue";
    AffirmationStatus["REJECTED"] = "rejected";
    AffirmationStatus["AFFIRMED"] = "affirmed";
    AffirmationStatus["FAILED"] = "failed";
    AffirmationStatus["CANCELED"] = "canceled";
})(AffirmationStatus = exports.AffirmationStatus || (exports.AffirmationStatus = {}));
class Affirmation {
    constructor(affirmationData, bitgo, enterpriseId) {
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.updateAffirmationData(affirmationData);
    }
    /**
     * Affirms a pending affirmation, authorizing the settlement
     * @param payload payload authorizing the movement of funds from a trading account
     * @param signature signature of the payload with the user key of the trading account
     */
    async affirm(payload, signature) {
        const body = {
            payload: JSON.stringify(payload),
            signature: signature,
        };
        return await this.updateStatus(AffirmationStatus.AFFIRMED, body);
    }
    /**
     * Rejects a pending affirmation, cancelling the settlement
     */
    async reject() {
        return this.updateStatus(AffirmationStatus.REJECTED, null);
    }
    /**
     * Cancels a pending affirmation, cancelling the settlement
     */
    async cancel() {
        return this.updateStatus(AffirmationStatus.CANCELED, null);
    }
    async updateStatus(status, body) {
        const bodyWithStatus = { status, ...body };
        const url = this.bitgo.microservicesUrl(`/api/trade/v1/enterprise/${this.enterpriseId}/account/${this.partyAccountId}/affirmations/${this.id}`);
        const response = await this.bitgo.put(url).send(bodyWithStatus).result();
        this.updateAffirmationData(response);
    }
    updateAffirmationData(affirmationData) {
        this.id = affirmationData.id;
        this.partyAccountId = affirmationData.partyAccountId;
        this.status = affirmationData.status;
        this.settlement = affirmationData.settlement;
        this.lock = affirmationData.lock;
        this.payload = affirmationData.payload;
        this.createdAt = new Date(affirmationData.createdAt);
        this.expireAt = new Date(affirmationData.expireAt);
    }
}
exports.Affirmation = Affirmation;
