// bitgo/dist/src/v2/trading/tradingPartners.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradingPartners = exports.TradingReferralRequesterSide = void 0;
const tradingPartner_1 = require("./tradingPartner");
// Side of the requester (if they should be considered the primary or the secondary)
// Only important for agency partnerships
// the primaryAccount is the agent, settling trades for the secondary account id
var TradingReferralRequesterSide;
(function (TradingReferralRequesterSide) {
    TradingReferralRequesterSide["PRIMARY"] = "primary";
    TradingReferralRequesterSide["SECONDARY"] = "secondary";
})(TradingReferralRequesterSide = exports.TradingReferralRequesterSide || (exports.TradingReferralRequesterSide = {}));
class TradingPartners {
    constructor(bitgo, enterpriseId, account) {
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.account = account;
    }
    /**
     * List all trading partners of your trading account. Your trading partners are the accounts you are allowed to settle with.
     */
    async list() {
        const url = this.bitgo.microservicesUrl(`/api/trade/v1/enterprise/${this.enterpriseId}/account/${this.account.id}/tradingpartners`);
        const response = (await this.bitgo.get(url).result());
        return response.tradingPartners.map((partner) => new tradingPartner_1.TradingPartner(partner, this.bitgo, this.enterpriseId, this.account));
    }
    /**
     * Add trading partner given the unique referralCode provided by trading partner.
     * @param params
     * @param params.referralCode unique referral code provided by counterparty
     * @param params.type type of trading partnership
     * @param params.requesterSide side of the requester (primary or secondary) important for agency relationships
     */
    async addByCode(params) {
        const url = this.bitgo.microservicesUrl(`/api/trade/v1/enterprise/${this.enterpriseId}/account/${this.account.id}/tradingpartners`);
        return await this.bitgo.post(url).send(params).result();
    }
}
exports.TradingPartners = TradingPartners;
