// bitgo/dist/src/v2/trading/tradingPartner.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradingPartner = exports.TradingPartnerType = exports.TradingPartnerStatus = void 0;
var TradingPartnerStatus;
(function (TradingPartnerStatus) {
    TradingPartnerStatus["ACCEPTED"] = "accepted";
    TradingPartnerStatus["REJECTED"] = "rejected";
    TradingPartnerStatus["CANCELED"] = "canceled";
    TradingPartnerStatus["PENDING"] = "pending";
})(TradingPartnerStatus = exports.TradingPartnerStatus || (exports.TradingPartnerStatus = {}));
// Type of partnership used for settlements
var TradingPartnerType;
(function (TradingPartnerType) {
    TradingPartnerType["DIRECT"] = "direct";
    TradingPartnerType["AGENCY"] = "agency";
})(TradingPartnerType = exports.TradingPartnerType || (exports.TradingPartnerType = {}));
class TradingPartner {
    constructor(tradingPartnerData, bitgo, enterpriseId, currentAccount) {
        this.id = tradingPartnerData.id;
        this.primaryEnterpriseName = tradingPartnerData.primaryEnterpriseName;
        this.primaryAccountId = tradingPartnerData.primaryAccountId;
        this.secondaryEnterpriseName = tradingPartnerData.secondaryEnterpriseName;
        this.secondaryAccountId = tradingPartnerData.secondaryAccountId;
        this.requesterAccountId = tradingPartnerData.requesterAccountId;
        this.status = tradingPartnerData.status;
        this.type = tradingPartnerData.type;
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.currentAccount = currentAccount;
    }
    /**
     * Check if a trading partner has enough funds to cover the cost of a trade.
     * @param currency the currency to check
     * @param amount the amount of currency to check, represented in base units (such as cents, satoshi, or wei)
     */
    async checkBalance(currency, amount) {
        const partnerAccountId = this.primaryAccountId === this.currentAccount.id ? this.secondaryAccountId : this.primaryAccountId;
        const url = this.bitgo.microservicesUrl(`/api/trade/v1/enterprise/${this.enterpriseId}/account/${this.currentAccount.id}/tradingpartners/${partnerAccountId}/balance`);
        const response = (await this.bitgo.get(url).query({ currency, amount }).result());
        return response.check;
    }
}
exports.TradingPartner = TradingPartner;
